{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**This iPython notebook consists of analysis scripts associated with the article  \"Effects of sequence motifs in the yeast 3′ untranslated region determined from massively-parallel assays of random sequences\" by Andrew Savinov, Benjamin M. Brandsen, Brooke E. Angell, Josh T. Cuperus, and Stanley Fields. Please cite this article (see below) if you make use of these scripts.**\n",
    "\n",
    "Full citation:\n",
    "<br/>Savinov A, Brandsen BM, Angell BE, Cuperus JT, Fields S (2021) Effects of sequence motifs in the yeast 3′ untranslated region determined from massively-parallel assays of random sequences. *bioRxiv*. doi: https://doi.org/10.1101/2021.03.27.437361.\n",
    "\n",
    "Input data files for the analyses below can be found in the associated Source Data, located in this figshare repository:\nhttps://figshare.com/articles/dataset/Source_Data_for_Savinov_et_al_2021_3_UTRs/16664143\ndoi: 10.6084/m9.figshare.16664143\n",
    "\n",
    "This code was written by Andrew Savinov.\n",
    "\n",
    "Last updated: July 27, 2021.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pandas\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import Levenshtein as lev\n",
    "import Bio.SeqUtils\n",
    "from Bio.SeqUtils import GC\n",
    "import itertools as itertools\n",
    "import scipy.stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load N50C library data \n",
    "N50C_df_path = \"[[DATA_DIRECTORY]]/\"\\\n",
    "+ \"N50C_data\" + \".csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df = pandas.DataFrame.from_csv(N50C_df_path, index_col=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df['rel_log2_score(rescaled+mean-norm)'] =\\\n",
    "(results_df['log2_score']+abs(results_df['log2_score'].min()))\\\n",
    " .div( (results_df['log2_score']+abs(results_df['log2_score'].min())).mean() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print results_df['log2_score'].min()\n",
    "print results_df['log2_score'].mean()\n",
    "print (results_df['log2_score']+abs(results_df['log2_score'].min())).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## to look at native yeast 3' UTR sequences and prior measurements of their properties\n",
    "nativeUTRs_df_path = \"[[DATA_DIRECTORY]]/\"\\\n",
    "+ \"Geisberg2014_native-mRNA-half-life_with_sequence.csv\" \n",
    "\n",
    "nativeUTRs_df = pandas.DataFrame.from_csv(nativeUTRs_df_path, index_col=None)\n",
    "nativeUTRs_df['UTR'] = nativeUTRs_df['sequence'].astype(str)\n",
    "nativeUTRs_df['Geisberg/min'] = nativeUTRs_df['Half-Life']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pandas.set_option('max_colwidth', 500)\n",
    "#pandas.reset_option('max_colwidth')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nativeUTRs_df['rel_Geisberg/min(mean-norm)'] = nativeUTRs_df['Geisberg/min']\\\n",
    ".div(nativeUTRs_df['Geisberg/min'].mean())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make FASTA file from dataframe of interest.\n",
    "# (FASTA file of N50C library sequences is needed for running FIMO analysis, see below)\n",
    "\n",
    "### pick one of the below options: #######\n",
    "# (1.) N50C library\n",
    "UTRs_df = results_df['UTR'] #select this for N50C library sequences (>=5 input reads, not pseudocounted)# \n",
    "\n",
    "## (2.) Native yeast 3' UTRs\n",
    "## UTRs_df = nativeUTRs_df['UTR'] #select this for native yeast sequences 3'UTR data\n",
    "#########################################\n",
    "\n",
    "length = UTRs_df.size\n",
    "counter = 0\n",
    "\n",
    "### pick one of the below options (select as appropriate matching the above selection) #\n",
    "# (1.)  N50C version:\n",
    "FASTA_path = \"[[DATA_DIRECTORY]]/\"\\\n",
    "+ \"N50C_data\" + \".fasta\"\n",
    "    \n",
    "## (2.) Native Yeast 3'UTR seqs version #\n",
    "## FASTA_path = \"[[DATA_DIRECTORY]]/\"\\\n",
    "## + \"Native_seqs\" + \".fasta\" \n",
    "#######################################################################################\n",
    "\n",
    "UTRs_df_FASTA = open(FASTA_path,\"w+\")\n",
    "\n",
    "for i in range(0,length):\n",
    "    UTRs_df_FASTA.write(\">\" + str(counter) + \"\\n\")\n",
    "    UTRs_df_FASTA.write(UTRs_df[counter] + \"\\n\")\n",
    "    counter+=1\n",
    "    \n",
    "UTRs_df_FASTA.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### set definition of enrichment_df (and add 'seq_ID' column) for certain downstream analyses\n",
    "\n",
    "enrichment_df = results_df.copy()  #>=5 input reads, not pseudocounted, N50C library\n",
    "\n",
    "enrichment_df['seq_ID']=enrichment_df.index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## define additional / alternative dataframe name for the N50C data, for easy reference, etc; used extensively in the below scripts\n",
    "N50C_df = results_df.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load N50EPC library growth selection dataframe:\n",
    "\n",
    "N50EPC_enr_df_path = \"[[DATA_DIRECTORY]]/\"\\\n",
    "+ \"N50EPC_data.csv\" \n",
    "N50EPC_enr_df = pandas.DataFrame.from_csv(N50EPC_enr_df_path, index_col=None)\n",
    "N50EPC_enr_df['seq_ID'] = N50EPC_enr_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## define additional / alternative dataframe name for the native UTRs half-life data; used extensively in the below scripts\n",
    "enrichment_df_native = nativeUTRs_df.copy() # Native yeast 3'UTRs dataframe\n",
    "enrichment_df_native['seq_ID']=enrichment_df_native.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## define 'UTRs_df' as desired (optional)\n",
    "#UTRs_df = results_df['UTR'] ## to look at N50C library (non-pseudocounted, >=5 input reads)\n",
    "UTRs_df = nativeUTRs_df['UTR'] ## to look at native yeast 3' UTRs data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CYC1_wt_N50 = str.upper('tcgagtcatgtaattagttatgtcacgcttacattcacgccctcccccca')\n",
    "print CYC1_wt_N50\n",
    "print ('N50C native CYC1 3\\' UTR variant Enr = '+str(N50C_df[N50C_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]))\n",
    "print ('N50EPC native CYC1 3\\' UTR variant Enr = '+str(N50EPC_enr_df[N50EPC_enr_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]))\n",
    "print\n",
    "print ('N50C mean Enr = '+str(np.mean(N50C_df['log2_score'])))\n",
    "print ('N50EPC mean Enr = '+str(np.mean(N50EPC_enr_df['log2_score'])))\n",
    "print\n",
    "N50C_mean_Enr_CYC1_norm = np.mean(N50C_df['log2_score']) - N50C_df[N50C_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "N50EPC_mean_Enr_CYC1_norm = np.mean(N50EPC_enr_df['log2_score']) - N50EPC_enr_df[N50EPC_enr_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "print ('N50C mean Enr, CYC1 normalized = '+str(N50C_mean_Enr_CYC1_norm))\n",
    "print ('N50EPC mean Enr, CYC1 normalized = '+str(N50EPC_mean_Enr_CYC1_norm))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df['log2_score_CYC1-norm'] = N50C_df['log2_score'] -\\\n",
    "   N50C_df[N50C_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "N50C_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50EPC_enr_df['log2_score_CYC1-norm'] = N50EPC_enr_df['log2_score'] -\\\n",
    "   N50EPC_enr_df[N50EPC_enr_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "N50EPC_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print ('N50C min Enr = '+str(np.min(N50C_df['log2_score'])))\n",
    "print ('N50C mean Enr = '+str(np.mean(N50C_df['log2_score'])))\n",
    "print ('N50C max Enr = '+str(np.max(N50C_df['log2_score'])))\n",
    "print\n",
    "print ('N50EPC min Enr = '+str(np.min(N50EPC_enr_df['log2_score'])))\n",
    "print ('N50EPC mean Enr = '+str(np.mean(N50EPC_enr_df['log2_score'])))\n",
    "print ('N50EPC max Enr = '+str(np.max(N50EPC_enr_df['log2_score'])))\n",
    "print\n",
    "print\n",
    "print ('N50C mean CYC1-norm. Enr = '+str(np.mean(N50C_df['log2_score_CYC1-norm'])))\n",
    "print ('N50EPC mean CYC1-norm. Enr = '+str(np.mean(N50EPC_enr_df['log2_score_CYC1-norm'])))\n",
    "print\n",
    "print ('N50C min CYC1-norm. Enr = '+str(np.min(N50C_df['log2_score_CYC1-norm'])))\n",
    "print ('N50C mean CYC1-norm. Enr = '+str(np.mean(N50C_df['log2_score_CYC1-norm'])))\n",
    "print ('N50C max CYC1-norm. Enr = '+str(np.max(N50C_df['log2_score_CYC1-norm'])))\n",
    "print\n",
    "print ('N50EPC min CYC1-norm. Enr = '+str(np.min(N50EPC_enr_df['log2_score_CYC1-norm'])))\n",
    "print ('N50EPC mean CYC1-norm. Enr = '+str(np.mean(N50EPC_enr_df['log2_score_CYC1-norm'])))\n",
    "print ('N50EPC max CYC1-norm. Enr = '+str(np.max(N50EPC_enr_df['log2_score_CYC1-norm'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######  N50C vs. N50EPC analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def regex_search_start(x):\n",
    "    try:\n",
    "        return re.search('AATAAA|AAAAAA', str(x)).start()\n",
    "    except:\n",
    "        return np.nan\n",
    "    \n",
    "def add_columns(analysisDF):\n",
    "    analysisDF['(TA)3_pos(first)'] = analysisDF['UTR'].str.find('TATATA')\n",
    "    analysisDF['AAWAAA_pos(first)'] = analysisDF['UTR'].apply\\\n",
    "    (lambda x: regex_search_start(x))\n",
    "    analysisDF['GC%'] = analysisDF['UTR'].apply(GC)\n",
    "    analysisDF['AT%'] = 100 - analysisDF['GC%']\n",
    "\n",
    "def sem(df):\n",
    "    return np.std(df) / np.sqrt(df.shape[0])\n",
    "\n",
    "def print_parameters(analysisDF,name):\n",
    "    print\n",
    "    print name + ':'\n",
    "    print 'N = ' + str(analysisDF.shape[0])\n",
    "    print 'mean +/- sem = ' + str(np.mean(analysisDF['log2_score'])) + ' +\\- ' + str(sem(analysisDF['log2_score']))\n",
    "    \n",
    "analysisDF = enrichment_df; add_columns(analysisDF)\n",
    "print_parameters(analysisDF,'N50C, all')\n",
    "analysisDF = enrichment_df[enrichment_df['UTR'].str.contains('TATATA')]; add_columns(analysisDF)\n",
    "print_parameters(analysisDF,'N50C, (TA)3-containing')\n",
    "bins = np.linspace(-10,10,20)\n",
    "plt.hist(enrichment_df['log2_score'], bins, alpha=0.5 ,color = 'blue', edgecolor='black',\\\n",
    "         density=True, label='N50C, all')\n",
    "plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'red', edgecolor='black',\\\n",
    "         density=True, label='N50C cont. (TA)3')\n",
    "plt.legend(loc='upper left')\n",
    "plt.show()\n",
    "\n",
    "analysisDF = N50EPC_enr_df; add_columns(analysisDF)\n",
    "print_parameters(analysisDF,'N50EPC, all')\n",
    "analysisDF = N50EPC_enr_df[N50EPC_enr_df['UTR'].str.contains('TATATA')]; add_columns(analysisDF)\n",
    "print_parameters(analysisDF,'N50EPC, (TA)3-containing')\n",
    "bins = np.linspace(-10,10,20)\n",
    "plt.hist(N50EPC_enr_df['log2_score'], bins, alpha=0.5 ,color = 'blue', edgecolor='black',\\\n",
    "         density=True, label='N50EPC, all')\n",
    "plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'red', edgecolor='black',\\\n",
    "         density=True, label='N50EPC cont. (TA)3')\n",
    "plt.legend(loc='upper left')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different versions of same motif (comparing small deviations from consensus)\n",
    "## Here: the TATATA motif\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "def list_params_avg_motif_effects__no_exclusion(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "\n",
    "        \n",
    "## consensus excluded from deviant versions\n",
    "print 'Plot with TATATA sequences excluded from analyses of alternative versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATA',\\\n",
    "               'AATATA','TTTATA','TAAATA','TATTTA','TATAAA','TATATT','AAATTT','TTTAAA','AAAAAA','TTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('TATATA motif (for alternative versions, \\nexclude sequences also containing TATATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n",
    "\n",
    "## consensus NOT excluded from deviant versions\n",
    "print 'Plot with TATATA sequences NOT excluded from analyses of alternative motif versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATA',\\\n",
    "               'AATATA','TTTATA','TAAATA','TATTTA','TATAAA','TATATT','AAATTT','TTTAAA','AAAAAA','TTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__no_exclusion(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('TATATA motif (for alternative versions, \\ndo NOT exclude sequences also containing TATATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different versions of same motif (comparing small deviations from consensus)\n",
    "## Here: the ATATAT motif [as comparison to TATATA, see above]\n",
    "\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains('TATATA')==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                                & (analysisDF['UTR'].str.contains('TATATA')==False)\\\n",
    "                                & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "def list_params_avg_motif_effects__no_exclusion(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains('TATATA')==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains('TATATA')==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus_but_not_TA3(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                                & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "## consensus excluded from deviant versions\n",
    "print 'Plot with ATATAT sequences excluded from analyses of alternative versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','ATATAT',\\\n",
    "               'TTATAT','AAATAT','ATTTAT','ATAAAT','ATATTT','ATATAA','AAATTT','TTTAAA','AAAAAA','TTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('ATATAT motif (exclude seqs also cont. TATATA from all);'\\\n",
    "          +'\\nfor alt. versions, *exclude* seqs also containing ATATAT)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n",
    "\n",
    "## consensus NOT excluded from deviant versions\n",
    "print 'Plot with ATATAT sequences NOT excluded from analyses of alternative motif versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','ATATAT',\\\n",
    "               'TTATAT','AAATAT','ATTTAT','ATAAAT','ATATTT','ATATAA','AAATTT','TTTAAA','AAAAAA','TTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__no_exclusion(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('ATATAT motif (exclude seqs also cont. TATATA from all);'\\\n",
    "          +'\\nfor alt. versions, do NOT exclude seqs also containing ATATAT)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different versions of same motif (comparing small deviations from consensus)\n",
    "## Here: the TATATA motif, replacing A or T with (A-->G, T-->C)\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "def list_params_avg_motif_effects__no_exclusion(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "\n",
    "        \n",
    "## consensus excluded from deviant versions\n",
    "print 'Plot with TATATA sequences excluded from analyses of alternative versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATA',\\\n",
    "               'CATATA','TGTATA','TACATA','TATGTA','TATACA','TATATG','AAATTT','TTTAAA','AAAAAA','TTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('TATATA motif and T->C, A->G subs at each site (for alt. versions, \\nexclude sequences also containing TATATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n",
    "\n",
    "## consensus NOT excluded from deviant versions\n",
    "print 'Plot with TATATA sequences NOT excluded from analyses of alternative motif versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATA',\\\n",
    "               'CATATA','TGTATA','TACATA','TATGTA','TATACA','TATATG','AAATTT','TTTAAA','AAAAAA','TTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__no_exclusion(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('TATATA motif and T->C, A->G subs at each site (for alt. versions, \\ndo NOT exclude sequences also containing TATATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different versions of same motif (comparing small deviations from consensus)\n",
    "## Here: the TATATATA, aka (TA)4, motif\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "def list_params_avg_motif_effects__no_exclusion(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "\n",
    "        \n",
    "## consensus excluded from deviant versions\n",
    "print 'Plot with (TA)4 sequences excluded from analyses of alternative versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATATA',\\\n",
    "               'AATATATA','TTTATATA','TAAATATA','TATTTATA','TATAAATA','TATATTTA','TATATAAA','TATATATT',\\\n",
    "               'AAAATTTT','TTTTAAAA','AAAAAAAA','TTTTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('(TA)4 motif (for alternative versions, \\nexclude sequences also containing TATATATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n",
    "\n",
    "## consensus NOT excluded from deviant versions\n",
    "print 'Plot with (TA)4 sequences NOT excluded from analyses of alternative motif versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATATA',\\\n",
    "               'AATATATA','TTTATATA','TAAATATA','TATTTATA','TATAAATA','TATATTTA','TATATAAA','TATATATT',\\\n",
    "               'AAAATTTT','TTTTAAAA','AAAAAAAA','TTTTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__no_exclusion(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('(TA)4 motif (for alternative versions, \\ndo NOT exclude sequences also containing TATATATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different versions of same motif (comparing small deviations from consensus)\n",
    "## Here: the TATATATATA, aka (TA)5, motif\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "def list_params_avg_motif_effects__no_exclusion(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "\n",
    "        \n",
    "## consensus excluded from deviant versions\n",
    "print 'Plot with (TA)5 sequences excluded from analyses of alternative versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATATATA',\\\n",
    "               'AATATATATA','TTTATATATA','TAAATATATA','TATTTATATA','TATAAATATA','TATATTTATA','TATATAAATA','TATATATTTA',\\\n",
    "               'TATATATAAA','TATATATATT','AAAAATTTTT','TTTTTAAAAA','AAAAAAAAAA','TTTTTTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue','blue','blue',\\\n",
    "                   'blue','blue','blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3.5)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('(TA)5 motif (for alternative versions, \\nexclude sequences also containing (TA)5)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n",
    "\n",
    "## consensus NOT excluded from deviant versions\n",
    "print 'Plot with (TA)5 sequences NOT excluded from analyses of alternative motif versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TATATATATA',\\\n",
    "               'AATATATATA','TTTATATATA','TAAATATATA','TATTTATATA','TATAAATATA','TATATTTATA','TATATAAATA','TATATATTTA',\\\n",
    "               'TATATATAAA','TATATATATT','AAAAATTTTT','TTTTTAAAAA','AAAAAAAAAA','TTTTTTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__no_exclusion(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue','blue','blue',\\\n",
    "                   'blue','blue','blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3.5)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('(TA)5 motif (for alternative versions, \\ndo NOT exclude sequences also containing (TA)5)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figures for motif effects: absolute and relative effects of each motif on random 50-mer library enrichment & native mRNA half-life ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: (TA)3 efficiency element\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TATATA'\n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 10\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "shuffles = shuffleList[1:]  # to exclude the consensus motif itself\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all NOT including motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains('TATATA')==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, TA3, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains('TATATA'))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                       ('rel_log2_score, TA3_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains('TATATA')==False)\\\n",
    "                                            & (enrichment_df['UTR'].str.contains('|'.join(shuffles)))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "                    \n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "        \n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all NOT including motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains('TATATA')==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), TA3',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains('TATATA')]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), TA3_shuffles',\\\n",
    "                        enrichment_df_native[(enrichment_df_native['UTR'].str.contains('TATATA')==False)\\\n",
    "                                            & (enrichment_df_native['UTR'].str.contains('|'.join(shuffles)))]\\\n",
    "                       [native_dataset]) ]\n",
    "\n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log2-enrichment)\n",
    "## Here: (TA)3 efficiency element \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TATATA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 10\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n",
    " \n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "shuffles = shuffleList[1:]  # to exclude the consensus motif itself\n",
    "\n",
    "dataframes_list = [ ('log2_score, all NOT including motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains('TATATA')==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, TA3, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains('TATATA'))]\\\n",
    "                       ['log2_score']),\\\n",
    "                       ('log2_score, TA3_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains('TATATA')==False)\\\n",
    "                                            & (enrichment_df['UTR'].str.contains('|'.join(shuffles)))]\\\n",
    "                       ['log2_score']) ]\n",
    "                    \n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "    \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: T5ATA efficiency element\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TTTTTATA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 15\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "shuffles = shuffleList[1:]  # to exclude the consensus motif itself\n",
    "\n",
    "\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all seqs WITHOUT T5ATA, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, T5ATA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, T5ATA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "                   \n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT T5ATA',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), T5ATA',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), T5ATA_shuffles',\\\n",
    "                    enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                        & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))\\\n",
    "                                        ]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "    \n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: T5ATA efficiency element\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TTTTTATA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 15\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "shuffles = shuffleList[1:]  # to exclude the consensus motif itself\n",
    "\n",
    "\n",
    "\n",
    "dataframes_list = [ ('log2_score, all seqs WITHOUT T5ATA, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, T5ATA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, T5ATA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "                   \n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different versions of same motif (comparing small deviations from consensus)\n",
    "## Here: the T5ATA motif\n",
    "\n",
    "def list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[(analysisDF['UTR'].str.contains(motif))\\\n",
    "                               & (analysisDF['UTR'].str.contains(sequence_list[1])==False)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + ', consensus version NOT present; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "        \n",
    "def list_params_avg_motif_effects__no_exclusion(analysisDF,variable,sequence_list):\n",
    "    i=0\n",
    "    for motif in sequence_list:\n",
    "        if i<2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        if i>=2:\n",
    "            testDF = analysisDF[analysisDF['UTR'].str.contains(motif)][variable]\n",
    "            #print 'properties of sequences containing motif ' + motif + '; mean, sem, and N:'\n",
    "        means_list.append(testDF.mean())\n",
    "        #print testDF.mean()\n",
    "        sems_list.append(testDF.sem())\n",
    "        #print testDF.sem()\n",
    "        N_list.append(testDF.shape[0])\n",
    "        #print testDF.shape[0]\n",
    "        names_list.append(sequence_list[i])\n",
    "        #print\n",
    "        i=i+1\n",
    "\n",
    "        \n",
    "## consensus excluded from deviant versions\n",
    "print 'Plot with T5ATA sequences excluded from analyses of alternative versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TTTTTATA',\\\n",
    "               'ATTTTATA','TATTTATA','TTATTATA','TTTATATA','TTTTAATA','TTTTTTTA','TTTTTAAA','TTTTTATT',\\\n",
    "               'AATTTTTT','TTTTTTAA','AAAAAAAA','TTTTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__exclude_consensus(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('T5ATA motif (for alternative versions, \\nexclude sequences also containing T5ATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print\n",
    "\n",
    "## consensus NOT excluded from deviant versions\n",
    "print 'Plot with T5ATA sequences NOT excluded from analyses of alternative motif versions:'\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "motifs_list = ('','TTTTTATA',\\\n",
    "               'ATTTTATA','TATTTATA','TTATTATA','TTTATATA','TTTTAATA','TTTTTTTA','TTTTTAAA','TTTTTATT',\\\n",
    "               'AATTTTTT','TTTTTTAA','AAAAAAAA','TTTTTTTT')\n",
    "analysisDF = results_df\n",
    "variable = 'log2_score'\n",
    "list_params_avg_motif_effects__no_exclusion(analysisDF,variable,motifs_list)\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue','blue',\\\n",
    "               'blue','blue','blue','blue','blue',\\\n",
    "                   'blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 3)\n",
    "plt.xlabel('motifs list index')\n",
    "plt.ylabel('average log2 enrichment')\n",
    "plt.title('T5ATA motif (for alternative versions, \\ndo NOT exclude sequences also containing T5ATA)')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print means_list\n",
    "print N_list\n",
    "print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: consensus positioning element, AAWAAA\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'AATAAA|AAAAAA'\n",
    "control_motif = 'AAATTT'\n",
    "\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, AAWAAA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, AAATTT, our data',\\\n",
    "                        enrichment_df[((enrichment_df['UTR'].str.contains(motif))==False) &\\\n",
    "                                        (enrichment_df['UTR'].str.contains(control_motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), alls eqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), AAWAAA',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), AAATTT',\\\n",
    "                    enrichment_df_native[((enrichment_df_native['UTR'].str.contains(motif))==False) &\\\n",
    "                                        (enrichment_df_native['UTR'].str.contains(control_motif))]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: consensus positioning element, AAWAAA\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'AATAAA|AAAAAA' \n",
    "control_motif = 'AAATTT'\n",
    "\n",
    "\n",
    "dataframes_list = [ ('log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, AAWAAA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, AAATTT, our data',\\\n",
    "                        enrichment_df[((enrichment_df['UTR'].str.contains(motif))==False) &\\\n",
    "                                        (enrichment_df['UTR'].str.contains(control_motif))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "        \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: GCGCGC control motif\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'GCGCGC' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 10\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "shuffles = shuffleList[1:]  # to exclude the consensus motif itself\n",
    "\n",
    "motif = 'GCGCGC'   \n",
    "\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all seqs WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, GCGCGC, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, GCGCGC shuffles, our data',\\\n",
    "                        enrichment_df[((enrichment_df['UTR'].str.contains(motif))==False) &\\\n",
    "                                        (enrichment_df['UTR'].str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), GCGCGC',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), GCGCGC shuffles',\\\n",
    "                    enrichment_df_native[((enrichment_df_native['UTR'].str.contains(motif))==False) &\\\n",
    "                                        (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "    #print names_list\n",
    "    #print means_list\n",
    "    #print sems_list\n",
    "    #print N_list\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: GCGCGC control motif\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'GCGCGC' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 10\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "shuffles = shuffleList[1:]  # to exclude the consensus motif itself\n",
    "\n",
    "motif = 'GCGCGC'   \n",
    "\n",
    "\n",
    "dataframes_list = [ ('log2_score, all seqs WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, GCGCGC, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, GCGCGC shuffles, our data',\\\n",
    "                        enrichment_df[((enrichment_df['UTR'].str.contains(motif))==False) &\\\n",
    "                                        (enrichment_df['UTR'].str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: Puf1/Puf2 binding site = UAAUNNNUAAU motif\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TAAT...TAAT' #'.' = regex term for any character (N) \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TAAT...TAAT'\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, TAATNNNTAAT, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, TAATNNNTAAT_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), TAATNNNTAAT',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), TAATNNNTAAT_shuffles',\\\n",
    "                    enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                        & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))\\\n",
    "                                        ]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: Puf1/Puf2 binding site = UAAUNNNUAAU motif\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TAAT...TAAT' #'.' = regex term for any character (N) \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TAAT...TAAT'\n",
    "\n",
    "dataframes_list = [ ('log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, TAATNNNTAAT, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, TAATNNNTAAT_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "## -- RELATIVE effects on E, native gene half-life\n",
    "## Here: Puf3 binding site motif = UGUANAUA \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TGTA.ATA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TGTA.ATA'\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, Puf3 consensus = TGTANATA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, TGTANATA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), Puf3 consensus = TGTANATA',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), TGTANATA_shuffles',\\\n",
    "                    enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                        & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))\\\n",
    "                                        ]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: Puf3 binding site motif = UGUANAUA \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TGTA.ATA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TGTA.ATA'\n",
    "\n",
    "dataframes_list = [ ('log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score Puf3 consensus = TGTANATA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, TGTANATA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: Puf4 binding site motif = UGUANANUA\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TGTA.A.TA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TGTA.A.TA'\n",
    "print motif\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, Puf4 consensus = TGTANANTA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, TGTANANTA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), Puf4 consensus = TGTANANTA',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), TGTANANTA_shuffles',\\\n",
    "                    enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                        & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))\\\n",
    "                                        ]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.6, 1.5)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: Puf4 binding site motif = UGUANANUA \n",
    "\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TGTA.A.TA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TGTA.A.TA'\n",
    "print motif\n",
    "\n",
    "dataframes_list = [ ('log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, Puf4 consensus = TGTANANTA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, TGTANANTA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Puf5 binding site motif = UGUANNNNUA \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TGTA....TA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TGTA....TA'\n",
    "print motif\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, Puf5 consensus = TGTANNNNTA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, TGTANNNNTA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), Puf5 consensus = TGTANNNNTA',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), TGTANNNNTA_shuffles',\\\n",
    "                    enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                        & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))\\\n",
    "                                        ]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Puf5 binding site motif = UGUANNNNUA \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TGTA....TA' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TGTA....TA'\n",
    "print motif\n",
    "\n",
    "dataframes_list = [ ('log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, Puf5 consensus = TGTANNNNTA, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, TGTANNNNTA_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: Puf6 binding site motif = UUGU \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TTGT' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 3\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TTGT'\n",
    "print motif\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, Puf6 site = TTGT, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, TTGT_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), Puf6 site = TTGT',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), TTGT_shuffles',\\\n",
    "                    enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                        & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))\\\n",
    "                                        ]\\\n",
    "                   [native_dataset])\\\n",
    "                      ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## Here: Puf6 binding site motif = UUGU \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TTGT' \n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 3\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "\n",
    "\n",
    "#print\n",
    "shuffles = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print shuffles\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TTGT'\n",
    "print motif\n",
    "\n",
    "dataframes_list = [ ('log2_score, all WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, Puf6 site = TTGT, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, TTGT_shuffles, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log-2 enrichment)\n",
    "## polyU_8X = UUUUUUUU;   first 25 nt, last 25 nt, ..., A4T4 shuffles control (5 bars total, see below)\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TTTTAAAA' # shuffle this to generate %AU control (4U's, 4 A's)\n",
    "\n",
    "minDist = len(motif)/2  ## this is in addition to extra filter to remove sequences with 'TATA' content (see below)\n",
    "print minDist\n",
    "\n",
    "numShuffles = 35\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList) & (('TATA' in shuffleX) == False)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "#set and print\n",
    "shuffles = shuffleList#[1:]  ## [1:] to exclude the consensus motif itself\n",
    "#print motif\n",
    "print shuffles\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "motif = 'TTTTTTTT'\n",
    "\n",
    "\n",
    "dataframes_list = [ ('log2_score, all seqs WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, polyU_8X, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, polyU_8X in first 25 nt, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str[1:25].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, polyU_8X in last 25 nt, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str[26:50].str.contains(motif))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, T4A4 control motif, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['log2_score']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- RELATIVE EFFECTS ON E, half-life\n",
    "## Here: polyU_8X = UUUUUUUU, with shuffles of T4A4 (inc T4A4) as control\n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = 'TTTTAAAA' # T4A4 shuffles control generation (for 4 U's + 4 A's control)\n",
    "\n",
    "minDist = len(motif)/2  ## this is in addition to extra filter to remove sequences with 'TATA' content (see below)\n",
    "print minDist\n",
    "\n",
    "numShuffles = 35\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList) & (('TATA' in shuffleX) == False)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "#set and print\n",
    "shuffles = shuffleList#[1:]  ## [1:] to exclude the consensus motif itself\n",
    "#print motif\n",
    "print shuffles\n",
    "\n",
    "print\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "\n",
    "motif = 'TTTTTTTT'\n",
    "controlMotif = 'TTAAATTT'\n",
    "print motif\n",
    "print controlMotif\n",
    "\n",
    "\n",
    "dataframes_list = [ ('rel_log2_score, all seqs WITHOUT motif, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                      ('rel_log2_score, polyU_8X, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']),\\\n",
    "                   ('rel_log2_score, T4A4 shuffles control, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       ['rel_log2_score(rescaled+mean-norm)']) ]\n",
    "\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "native_dataset_list = ['rel_Geisberg/min(mean-norm)']\n",
    "for native_dataset in native_dataset_list:\n",
    "\n",
    "    print native_dataset\n",
    "    \n",
    "    dataframes_list = [ ('rel_nativeUTRs_halflife(/min), all seqs WITHOUT motif',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)==False]\\\n",
    "                       [native_dataset]),\\\n",
    "                      ('rel_nativeUTRs_halflife(/min), polyU_8X',\\\n",
    "                       enrichment_df_native[enrichment_df_native['UTR'].str.contains(motif)]\\\n",
    "                       [native_dataset]),\\\n",
    "                       ('rel_nativeUTRs_halflife(/min), T4A4 shuffles control',\\\n",
    "                       enrichment_df_native[(enrichment_df_native['UTR'].str.contains(motif)==False)\\\n",
    "                                            & (enrichment_df_native['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles))))]\\\n",
    "                       [native_dataset]) ]\n",
    "    \n",
    "\n",
    "    for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue','green','green','green']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0.9, 1.3)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bar plots of different motifs, different datasets, for N50C Enr (E) data as well as native UTR half-life data;\n",
    "## plotting effects for one motif at a time (& the associated controls).\n",
    "#  -- ABSOLUTE EFFECTS ON E (log2 enr)\n",
    "## Here: polyU (8x) vs U-rich controls with no more than two sequential U nucleotides \n",
    "\n",
    "## make shuffles (using my existing shufflemaker script, reproduced here)\n",
    "\n",
    "motif = '....TTTTTTTT....' # NNNNUUUUUUUUNNNN polyU vs U-rich control (for 8 U's + 8 N's)\n",
    "\n",
    "minDist = len(motif)/2  ## this is in addition to new filter removing sequences with 'TTT' content (see below)\n",
    "print minDist\n",
    "\n",
    "numShuffles = 50\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)\\\n",
    "            & (('TTT' in shuffleX) == False)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "#print shuffleList\n",
    "\n",
    "#set and print: motif 1\n",
    "motif_1 = motif\n",
    "shuffles_1 = shuffleList[1:]  ## [1:] to exclude the consensus motif itself\n",
    "print motif_1\n",
    "print shuffles_1\n",
    "\n",
    "\n",
    "means_list = []\n",
    "sems_list = []\n",
    "names_list = []\n",
    "N_list = []\n",
    "\n",
    "dataframes_list = [ ('log2_score, seqs WITHOUT N4U8N4, our data',\\\n",
    "                         enrichment_df[(enrichment_df['UTR'].str.contains(motif_1)==False)]\\\n",
    "                       ['log2_score']),\\\n",
    "                      ('log2_score, seqs WITH N4U8N4, our data',\\\n",
    "                       enrichment_df[(enrichment_df['UTR'].str.contains(motif_1))]\\\n",
    "                       ['log2_score']),\\\n",
    "                   ('log2_score, seqs WITH SHUFFLES of N4U8N4 only, our data',\\\n",
    "                        enrichment_df[(enrichment_df['UTR'].str.contains(motif_1)==False)\\\n",
    "                                           & (enrichment_df['UTR'].str.contains('TATATA')==False)\\\n",
    "                                      & (enrichment_df['UTR']\\\n",
    "                                               .str.contains(('|'.join(shuffles_1))))]\\\n",
    "                       ['log2_score']) ]\n",
    "print\n",
    "\n",
    "\n",
    "for dataframe in dataframes_list:\n",
    "        means_list.append(dataframe[1].mean())\n",
    "        sems_list.append(dataframe[1].sem())\n",
    "        names_list.append(dataframe[0])\n",
    "        N_list.append(dataframe[1].shape[0])\n",
    "\n",
    "            \n",
    "\n",
    "bar_x = list(range(len(names_list)))\n",
    "#print bar_x\n",
    "colors_list = ['blue','blue','blue']\n",
    "\n",
    "edge_colors_list = []\n",
    "for member in bar_x:\n",
    "    edge_colors_list.append('black')\n",
    "\n",
    "plt.bar(x=bar_x, height=means_list, yerr=sems_list, capsize=4.0, color=colors_list,\\\n",
    "        edgecolor=edge_colors_list)#, tick_label=names_list)\n",
    "plt.ylim(0, 2.75)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print names_list\n",
    "print N_list\n",
    "print means_list\n",
    "print sems_list\n",
    "print\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## further analyses of (TA)3-containing sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df['GC%'] = N50C_df['UTR'].apply(GC)\n",
    "N50C_df['AT%'] = 100 - N50C_df['GC%']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df['(TA)3_pos(first)'] = N50C_df['UTR'].str.find('TATATA')\n",
    "#N50C_df['AAWAAA_pos(first)'] = N50C_df['UTR'].str.find('AATAAA|AAAAAA')\n",
    "\n",
    "def regex_search_start(x):\n",
    "    try:\n",
    "        return re.search('AATAAA|AAAAAA', str(x)).start()\n",
    "    except:\n",
    "        return np.nan\n",
    "\n",
    "N50C_df['AAWAAA_pos(first)'] = N50C_df['UTR'].apply\\\n",
    "(lambda x: regex_search_start(x))\n",
    "#(lambda x: (re.search('AATAAA|AAAAAA', str(x))))\n",
    "\n",
    "N50C_df.loc[N50C_df['(TA)3_pos(first)']==-1,'(TA)3_pos(first)']\\\n",
    "= np.nan\n",
    "\n",
    "N50C_df.loc[N50C_df['AAWAAA_pos(first)']==-1,'AAWAAA_pos(first)']\\\n",
    "= np.nan\n",
    "\n",
    "N50C_df.loc[N50C_df['(TA)3_pos(first)']>=0,'(TA)3_pos(first)']\\\n",
    "= N50C_df['(TA)3_pos(first)'] + 1\n",
    "\n",
    "N50C_df.loc[N50C_df['AAWAAA_pos(first)']>=0,'AAWAAA_pos(first)']\\\n",
    "= N50C_df['AAWAAA_pos(first)'] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print N50C_df.shape[0]\n",
    "N50C_df_w_TA3 = N50C_df[N50C_df['UTR'].str.contains('TATATA')]\n",
    "print N50C_df_w_TA3.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# analysis of N50C sequences with multiple occurences of the consensus efficiency element, UAUAUA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "N50C_df['(TA)3_num_instances'] = N50C_df['UTR'].str.count('TATATA')\n",
    "N50C_df_w_TA3 = N50C_df[N50C_df['UTR'].str.contains('TATATA')]\n",
    "N50C_df_w_mult_TA3_nonoverlapping = N50C_df[N50C_df['(TA)3_num_instances']>=2]\n",
    "\n",
    "print 'N50C_all'; print N50C_df.shape[0]; print np.mean(N50C_df['log2_score']); print sem(N50C_df['log2_score']); print\n",
    "print 'N50C_containing_TA3'; print N50C_df_w_TA3.shape[0]; print np.mean(N50C_df_w_TA3['log2_score']);\\\n",
    "      print sem(N50C_df_w_TA3['log2_score']); print\n",
    "print 'N50C_containing_2X_TA3(nonoverlapping)'; print N50C_df_w_mult_TA3_nonoverlapping.shape[0];\\\n",
    "         print np.mean(N50C_df_w_mult_TA3_nonoverlapping['log2_score']);\\\n",
    "        print sem(N50C_df_w_mult_TA3_nonoverlapping['log2_score']); print\n",
    "\n",
    "#print \n",
    "\n",
    "def make_shuffles(motif_input,numShuffles_input):\n",
    "    \n",
    "    motif = motif_input\n",
    "    numShuffles = numShuffles_input\n",
    "    \n",
    "    minDist = len(motif)/2\n",
    "    #print minDist\n",
    "\n",
    "    shuffleList = []\n",
    "    shuffleList.append(motif)\n",
    "    #print shuffleList\n",
    "\n",
    "    for i in range (0,numShuffles): #(0,5):\n",
    "        continueLoop = True\n",
    "        while(continueLoop == True):\n",
    "            shuffleX = motif\n",
    "            #print shuffleX\n",
    "            shuffleX_list = list(shuffleX)\n",
    "            random.shuffle(shuffleX_list)\n",
    "            shuffleX = ''.join(shuffleX_list)\n",
    "            #print shuffleX\n",
    "            dist = lev.hamming(shuffleX, motif)\n",
    "            if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "                #print \"success\"\n",
    "                shuffleList.append(shuffleX)\n",
    "                continueLoop = False\n",
    "            #else:\n",
    "                #print \"fail, try again\"\n",
    "\n",
    "    #print;print; print shuffleList\n",
    "    return shuffleList\n",
    "\n",
    "motif='TATATA'\n",
    "numShuffles = 10\n",
    "shuffleList = make_shuffles(motif,numShuffles)\n",
    "shuffles = shuffleList[1:]\n",
    "\n",
    "#print\n",
    "N50C_df_w_1X_TA3 = N50C_df[N50C_df['(TA)3_num_instances']==1]\n",
    "\n",
    "print 'N50C_containing_1X_TA3'; print N50C_df_w_1X_TA3.shape[0]; print np.mean(N50C_df_w_1X_TA3['log2_score']);\\\n",
    "      print sem(N50C_df_w_1X_TA3['log2_score']); print\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bin TA3-cont. sequences by enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df_w_TA3_bin1 = N50C_df_w_TA3[N50C_df_w_TA3['log2_score']<-3]\n",
    "print N50C_df_w_TA3_bin1.shape[0]\n",
    "\n",
    "N50C_df_w_TA3_bin2 = N50C_df_w_TA3[(N50C_df_w_TA3['log2_score']>=-3)\\\n",
    "                                   &(N50C_df_w_TA3['log2_score']<-1)]\n",
    "print N50C_df_w_TA3_bin2.shape[0]\n",
    "\n",
    "N50C_df_w_TA3_bin3 = N50C_df_w_TA3[(N50C_df_w_TA3['log2_score']>=-1)\\\n",
    "                                   &(N50C_df_w_TA3['log2_score']<1)]\n",
    "print N50C_df_w_TA3_bin3.shape[0]\n",
    "\n",
    "N50C_df_w_TA3_bin4 = N50C_df_w_TA3[(N50C_df_w_TA3['log2_score']>=1)\\\n",
    "                                   &(N50C_df_w_TA3['log2_score']<3)]\n",
    "print N50C_df_w_TA3_bin4.shape[0]\n",
    "\n",
    "N50C_df_w_TA3_bin5 = N50C_df_w_TA3[(N50C_df_w_TA3['log2_score']>=3)\\\n",
    "                                   &(N50C_df_w_TA3['log2_score']<6)]\n",
    "print N50C_df_w_TA3_bin5.shape[0]\n",
    "\n",
    "N50C_df_w_TA3_bin6 = N50C_df_w_TA3[(N50C_df_w_TA3['log2_score']>=6)\\\n",
    "                                   &(N50C_df_w_TA3['log2_score']<9)]\n",
    "print N50C_df_w_TA3_bin6.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df_bin1 = N50C_df[N50C_df['log2_score']<-3]\n",
    "print N50C_df_bin1.shape[0]\n",
    "\n",
    "N50C_df_bin2 = N50C_df[(N50C_df['log2_score']>=-3)\\\n",
    "                                   &(N50C_df['log2_score']<-1)]\n",
    "print N50C_df_bin2.shape[0]\n",
    "\n",
    "N50C_df_bin3 = N50C_df[(N50C_df['log2_score']>=-1)\\\n",
    "                                   &(N50C_df['log2_score']<1)]\n",
    "print N50C_df_bin3.shape[0]\n",
    "\n",
    "N50C_df_bin4 = N50C_df[(N50C_df['log2_score']>=1)\\\n",
    "                                   &(N50C_df['log2_score']<3)]\n",
    "print N50C_df_bin4.shape[0]\n",
    "\n",
    "N50C_df_bin5 = N50C_df[(N50C_df['log2_score']>=3)\\\n",
    "                                   &(N50C_df['log2_score']<6)]\n",
    "print N50C_df_bin5.shape[0]\n",
    "\n",
    "N50C_df_bin6 = N50C_df[(N50C_df['log2_score']>=6)\\\n",
    "                                   &(N50C_df['log2_score']<9)]\n",
    "print N50C_df_bin6.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df_w_TA3_and_AAWAAA = N50C_df_w_TA3[N50C_df_w_TA3['UTR'].str.contains('AATAAA|AAAAAA')]\n",
    "N50C_df_w_TA3_and_AAWAAA.head(3)\n",
    "pos_before_eff_df = N50C_df_w_TA3_and_AAWAAA[N50C_df_w_TA3_and_AAWAAA['(TA)3_pos(first)']\\\n",
    "                                      > N50C_df_w_TA3_and_AAWAAA['AAWAAA_pos(first)']]\n",
    "eff_before_pos_df = N50C_df_w_TA3_and_AAWAAA[N50C_df_w_TA3_and_AAWAAA['(TA)3_pos(first)']\\\n",
    "                                      < N50C_df_w_TA3_and_AAWAAA['AAWAAA_pos(first)']]\n",
    "print N50C_df_w_TA3_and_AAWAAA.shape[0]\n",
    "print pos_before_eff_df.shape[0]\n",
    "print eff_before_pos_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N50C_df.name = 'N50C_df'\n",
    "N50C_df_w_TA3.name = 'N50C_df_w_TA3'\n",
    "\n",
    "N50C_df_w_TA3_and_AAWAAA = N50C_df_w_TA3[N50C_df_w_TA3['UTR'].str.contains('AATAAA|AAAAAA')]\n",
    "N50C_df_w_TA3_and_AAWAAA.name = 'N50C_df_w_TA3_and_AAWAAA'\n",
    "\n",
    "pos_before_eff_df = N50C_df_w_TA3_and_AAWAAA[N50C_df_w_TA3_and_AAWAAA['(TA)3_pos(first)']\\\n",
    "                                      > N50C_df_w_TA3_and_AAWAAA['AAWAAA_pos(first)']]\n",
    "pos_before_eff_df.name = 'pos_before_eff_df'\n",
    "\n",
    "eff_before_pos_df = N50C_df_w_TA3_and_AAWAAA[N50C_df_w_TA3_and_AAWAAA['(TA)3_pos(first)']\\\n",
    "                                      < N50C_df_w_TA3_and_AAWAAA['AAWAAA_pos(first)']]\n",
    "eff_before_pos_df.name = 'eff_before_pos_df'\n",
    "\n",
    "print N50C_df.name + \" mean log2 score = \" + str((N50C_df['log2_score']).mean())\n",
    "print N50C_df_w_TA3.name + \" mean log2 score = \" + str((N50C_df_w_TA3['log2_score']).mean())\n",
    "print N50C_df_w_TA3_and_AAWAAA.name + \" mean log2 score = \" + str((N50C_df_w_TA3_and_AAWAAA['log2_score']).mean())\n",
    "print \"N50C with pos --> eff mean log2 score = \" + str((pos_before_eff_df['log2_score']).mean())\n",
    "print \"N50C with eff --> pos mean log2 score = \" + str((eff_before_pos_df['log2_score']).mean())\n",
    "print\n",
    "print\n",
    "\n",
    "\n",
    "dfs_list = [N50C_df_w_TA3_bin2, N50C_df_w_TA3_bin3, N50C_df_w_TA3_bin4,\\\n",
    "           N50C_df_w_TA3_bin5, N50C_df_w_TA3_bin6]   #exclude bin1 because N=2 for this bin\n",
    "\n",
    "N50C_df_w_TA3_bin2.name = 'N50C_df_w_TA3_bin2'\n",
    "N50C_df_w_TA3_bin3.name = 'N50C_df_w_TA3_bin3'\n",
    "N50C_df_w_TA3_bin4.name = 'N50C_df_w_TA3_bin4'\n",
    "N50C_df_w_TA3_bin5.name = 'N50C_df_w_TA3_bin5'\n",
    "N50C_df_w_TA3_bin6.name = 'N50C_df_w_TA3_bin6'\n",
    "\n",
    "bins_fraction_posBeforeEff_list = []\n",
    "bins_fraction_effBeforePos_list = []\n",
    "bins_count_posBeforeEff_list = []\n",
    "bins_count_effBeforePos_list = []\n",
    "bins_count_pos_and_eff_list = []\n",
    "bins_GC_mean_list = []\n",
    "bins_GC_sem_list = []\n",
    "bins_count_in_bin_list = []\n",
    "bins_AT_mean_list = []\n",
    "bins_AT_sem_list = []\n",
    "\n",
    "for analysisDF in dfs_list:\n",
    "    pos_and_eff_df = analysisDF[analysisDF['UTR'].str.contains('AAAAAA|AATAAA')]\n",
    "    pos_before_eff_df = pos_and_eff_df[pos_and_eff_df['(TA)3_pos(first)']\\\n",
    "                                      > pos_and_eff_df['AAWAAA_pos(first)']]\n",
    "    eff_before_pos_df = pos_and_eff_df[pos_and_eff_df['(TA)3_pos(first)']\\\n",
    "                                      < pos_and_eff_df['AAWAAA_pos(first)']]\n",
    "    number_pos_and_eff = float(pos_and_eff_df.shape[0])\n",
    "    number_pos_first = float(pos_before_eff_df.shape[0])\n",
    "    number_eff_first = float(eff_before_pos_df.shape[0])\n",
    "    fraction_pos_first = number_pos_first / number_pos_and_eff\n",
    "    fraction_eff_first = number_eff_first / number_pos_and_eff\n",
    "    print \"For the data subset \" + str(analysisDF.name) + \":\"\n",
    "    print \"whole-bin log2 score min = \" + str(analysisDF['log2_score'].min())\n",
    "    print \"whole-bin log2 score max = \" + str(analysisDF['log2_score'].max())\n",
    "    print \"number UTRs containing both pos and eff elements:  \" + str(number_pos_and_eff)\n",
    "    print \"number UTRs containing both, pos BEFORE eff (pos --> eff):  \" + str(number_pos_first)\n",
    "    print \"number UTRs containing both, pos AFTER eff (eff --> pos):  \" + str(number_eff_first)\n",
    "    print \"fraction with pos elem --> eff elem: \"\n",
    "    print fraction_pos_first\n",
    "    print \"fraction with eff elem --> pos elem:\"\n",
    "    print fraction_eff_first\n",
    "    print \"mean log2 score of entire bin = \" + str(analysisDF['log2_score'].mean())\n",
    "    print \"mean log2 score of bin sequences containing pos AND eff elems = \" + str(pos_and_eff_df['log2_score'].mean())\n",
    "    print \"mean log2 score of bin sequences containing pos --> eff = \" + str(pos_before_eff_df['log2_score'].mean())\n",
    "    print \"mean log2 score of bin sequences containing eff --> pos = \" + str(eff_before_pos_df['log2_score'].mean())\n",
    "    print\n",
    "    bins_fraction_posBeforeEff_list.append(fraction_pos_first)\n",
    "    bins_fraction_effBeforePos_list.append(fraction_eff_first)\n",
    "    bins_count_posBeforeEff_list.append(number_pos_first)\n",
    "    bins_count_effBeforePos_list.append(number_eff_first)\n",
    "    bins_count_pos_and_eff_list.append(number_pos_and_eff)\n",
    "    bins_GC_mean_list.append(analysisDF['GC%'].mean())\n",
    "    bins_GC_sem_list.append(analysisDF['GC%'].sem())\n",
    "    bins_count_in_bin_list.append(analysisDF.shape[0])\n",
    "    bins_AT_mean_list.append(analysisDF['AT%'].mean())\n",
    "    bins_AT_sem_list.append(analysisDF['AT%'].sem())\n",
    "    \n",
    "bar_x = ['[-3,-1)', '[-1,1)', '[1,3)', '[3,6)', '[6,9)']\n",
    "\n",
    "plt.bar(x=bar_x, height=bins_fraction_effBeforePos_list, color='blue',\\\n",
    "        edgecolor='black')#, tick_label=bins_count_effBeforePos_list)\n",
    "for i in range(len(bar_x)):\n",
    "    plt.annotate(str(bins_count_pos_and_eff_list[i]),xy=(bar_x[i],bins_fraction_effBeforePos_list[i]+0.01))\n",
    "\n",
    "plt.show()\n",
    "\n",
    "bar_x = ['[-3,-1)', '[-1,1)', '[1,3)', '[3,6)', '[6,9)']\n",
    "\n",
    "plt.bar(x=bar_x, height=bins_fraction_effBeforePos_list, color='blue',\\\n",
    "        edgecolor='black')#, tick_label=bins_count_effBeforePos_list)\n",
    "\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### positional analyses of motif effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### \"Shuffle-maker\" script: generate desired number of shuffled sequences derived from motif of interest. \n",
    "### The shuffles (and nonshuffled original motif) generated here are exported for downstream use\n",
    "### in finding all instances of the motif and its shuffled versions in the N50 sequences using FIMO (see below)\n",
    "\n",
    "### The example shown here is for U5AUA (alternative efficiency element)\n",
    "\n",
    "motif = 'TTTTTATA'    #N=15  #Efficiency element v2\n",
    "\n",
    "#'TATATA'      #N=10  #Consensus efficiency element\n",
    "#'TTTTTATA'    #N=15  #Efficiency element v2\n",
    "#'GCGCGC'      #N=10  #GC-rich control motif\n",
    "#'UAAUNNNUAAU' #N=50  #PUF1/PUF2 motif\n",
    "#'TGTANATA'    #N=50 #TGTANATA-Puf3-FINAL_CORRECT for MS\n",
    "#'TGTANANTA'   #N=50 #TGTANANTA-Puf4-FINAL_CORRECT for MS\n",
    "#'TGTANNNNTA'  #N=50 #TGTANNNNTA-Puf5-FINAL-CORRECT for MS\n",
    "#'TTGT'        #N=3  #TTGT-Puf6-FINAL_CORRECT\n",
    "\n",
    "minDist = len(motif)/2\n",
    "print minDist\n",
    "\n",
    "numShuffles = 15\n",
    "\n",
    "shuffleList = []\n",
    "shuffleList.append(motif)\n",
    "#print shuffleList\n",
    "\n",
    "for i in range (0,numShuffles): #(0,5):\n",
    "    while(True):\n",
    "        shuffleX = motif\n",
    "        #print shuffleX\n",
    "        shuffleX_list = list(shuffleX)\n",
    "        random.shuffle(shuffleX_list)\n",
    "        shuffleX = ''.join(shuffleX_list)\n",
    "        #print shuffleX\n",
    "        dist = lev.hamming(shuffleX, motif)\n",
    "        if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "            #print \"success\"\n",
    "            shuffleList.append(shuffleX)\n",
    "            break\n",
    "        #else:\n",
    "            #print \"fail, try again\"\n",
    "\n",
    "print shuffleList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_motif_shuffles_test = pandas.DataFrame({'motif':shuffleList})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make text file of starting motif & its shuffles\n",
    "\n",
    "length = len(shuffleList)\n",
    "counter = 0\n",
    "path = \"[[YOUR_PATH_HERE]]/\"\\\n",
    "+ \"efficiency_TTTTTATA_with_15_shuffles\" + \".txt\"\n",
    "\n",
    "#+ \"efficiency_TATATA_with_10_shuffles\" + \".txt\" #consensus efficiency element\n",
    "#+ \"efficiency_TTTTTATA_with_15_shuffles\" + \".txt\" #alternative efficiency element\n",
    "#+ \"control_GCGCGC_with_10_shuffles\" + \".txt\" #(GC)3 control motif\n",
    "#+ \"PUF1&2_UAAUNNNUAAU_with_50_shuffles\" + \".txt\" #Puf1/Puf2 binding motif\n",
    "#+ \"PUF3_TGTANATA_with_50_shuffles\" + \".txt\"  #Puf3 binding site\n",
    "#+ \"Puf4_TGTANANTA_with_50_shuffles.txt\"  #Puf4 binding site\n",
    "#+ \"Puf5_TGTANNNNTA_with_50_shuffles.txt\"  #Puf5 binding site\n",
    "#+ \"Puf6_TTGT_with_3_shuffles.txt\"  #Puf6 binding site\n",
    "\n",
    "\n",
    "df_motif_shuffles_txt = open(path,\"w+\")\n",
    "\n",
    "for i in range(0,length):\n",
    "    df_motif_shuffles_txt.write(shuffleList[counter] + \"\\n\" + \"\\n\")\n",
    "    counter+=1\n",
    "    \n",
    "df_motif_shuffles_txt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make a set of \".sites\" format files for the starting motif and each of its shuffles;\n",
    "## these files are used as input for running the MEME Suite program FIMO in the command line, see below\n",
    "\n",
    "length = len(shuffleList)\n",
    "counter = 0\n",
    "motifName = 'efficiency_v2_TTTTTATA'      #alternative efficiency element\n",
    "## ** Also need to pre-generate a folder corresponding to the name of the motif in directory 'direc', see below! **\n",
    "\n",
    "#'efficiency_consensus_TATATA'  #consensus efficiency element\n",
    "#'efficiency_v2_TTTTTATA'       #alternative efficiency element\n",
    "#'control_GCGCGC'               #(GC)3 control motif'\n",
    "#'PUF1&2_UAAUNNNUAAU'           #Puf1/Puf2 binding site motif \n",
    "#'Puf3_TGTANATA'                #Puf3 binding site motif\n",
    "#'Puf4_TGTANANTA'               #Puf4 binding site motif\n",
    "#'Puf5_TGTANNNNTA'              #Puf5 binding site motif\n",
    "#'Puf6_TTGT'                    #Puf6 binding site motif\n",
    "\n",
    "\n",
    "motifID = motifName + \".sites\"\n",
    "direc = \"[[YOUR_PATH_HERE]]/\" + motifName + \"/\"  ## remember to pre-generate the \"motifName\" folder\n",
    "\n",
    "for i in range(0,length):\n",
    "    if counter==0:\n",
    "        path = direc + motifID\n",
    "    else:\n",
    "        motifID = motifName + \"_shuffle\" + str(counter) + \".sites\" \n",
    "        path = direc + motifID\n",
    "        \n",
    "    df_motif_shuffles_sites = open(path,\"w+\")\n",
    "\n",
    "    df_motif_shuffles_sites.write(shuffleList[counter])\n",
    "    \n",
    "    counter+=1\n",
    "    \n",
    "    df_motif_shuffles_sites.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Script for running sites2meme and FIMO from the MEME Suite software package\n",
    "### in order to find *all perfect matches* to the motif and each of its shuffled sequences\n",
    "### in the set of random 50-mer sequences interrogated by our high-throughput assay.\n",
    "\n",
    "## NOTE that this is script is run in the command line using a local installation of the MEME Suite programs;\n",
    "## will not run in iPython / Jupyter. For this reason the script is provided here as a triple-quoted literal\n",
    "## that can be copied to run it in the command line.\n",
    "\n",
    "# Example provided for the case of U5AUA motif (alt. efficiency element):\n",
    "\n",
    "'''\n",
    "\n",
    "#Running sites2meme to convert \".sites\" files to MEME format for FIMO analysis:\n",
    "sites2meme -ext .sites [[YOUR_PATH_HERE]]/efficiency_v2_TTTTTATA/ -> [[SITES2MEME_OUTPUT_DIR]]/efficiency_v2_TTTTTATA_w_shuffles_meme.txt &\n",
    "\n",
    "#Running FIMO to find all perfect matches to motif sequence and shuffled sequences:\n",
    "fimo --norc --thresh 1.6e-5 --bfile --uniform-- --oc [[FIMO_OUTPUT_DIR]]/efficiency_v2_TTTTTATA_w_shuffles_FIMOoutput [[SITES2MEME_OUTPUT_DIR]]/efficiency_v2_TTTTTATA_w_shuffles_meme.txt [[DATA_DIRECTORY]]/N50C_data.fasta &\n",
    "\n",
    "## notes: \n",
    "## --norc --> no reverse complement, search given strand only\n",
    "## --thresh --> set minimum p-value for matching. If uniform base probabilities, P-value of <1.5e-5 required for 8-residue motif ((0.25)^8)\n",
    "## --o --> Create a folder called dir and write output files in it. This option is not compatible with --oc as only one output folder is allowed.\n",
    "## --oc --> Create a folder called dir but if it already exists allow overwriting the contents. This option is not compatible with --o as only one output folder is allowed.\n",
    "## --bfile --> Specify the source of a 0-order background model for converting a frequency matrix to a log-odds score matrix and for use in estimating the p-values of match scores. The background model normalizes for biased distribution of individual letters in the sequences. The value of file is either the path to a file in Markov Background Model Format, or one of the keywords --motif--, motif-file or --uniform--. The first two keywords cause the 0-order letter frequencies contained in the motif file to be used, and --uniform-- causes uniform letter frequencies to be used. If the background model in file is higher than 0-order, only the 0-order portion is used. If both strands are being scored, the background model is modified by averaging the frequencies of letters and their reverse complements.\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FIMO_results_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min_df_path =\\\n",
    "\"[[FIMO_OUTPUT_PATH]]/\"\\\n",
    "+ \"efficiency_v2_TTTTTATA_w_shuffles_FIMOoutput/\"\\\n",
    "+ \"fimo\" + \".tsv\"\n",
    "\n",
    "FIMO_results_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min =\\\n",
    "pandas.read_csv(FIMO_results_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min_df_path,sep=None)\n",
    "\n",
    "FIMO_results_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min.rename(columns={'sequence_name': 'seq_ID'}, inplace=True)\n",
    "\n",
    "FIMO_results_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min.sort_values(by='motif_id').head(7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min = pandas.merge\\\n",
    "(FIMO_results_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min, enrichment_df, on='seq_ID')\n",
    "enrichment_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min\\\n",
    "[enrichment_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min['motif_id']=='efficiency_v2_TTTTTATA']\\\n",
    ".sort_values(by='start').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Script for the generation of positional mean enrichment plots for motif and average of shuffled motif control sequences\n",
    "# example below for alternative efficiency element, UUUUUAUA (U5AUA):\n",
    "\n",
    "motifName = 'efficiency_v2_TTTTTATA' #15 shuffles\n",
    "\n",
    "dataframe = enrichment_T5ATA_efficiency_v2_w_shuffles_N50C_noPC_5min\n",
    "\n",
    "counter = 0      #initialize counter (see below)\n",
    "lastMotif = 15   # lastMotif = -1 \n",
    "# use lastMotif = -1 to cancel shuffle plot calculations and plotting, only plot motif positional effects (e.g. for AAWAAA)\n",
    "# otherwise, lastMotif should match the # of shuffles for the motif being analyzed (e.g., 15 for U5AUA)\n",
    "end = lastMotif+1\n",
    "\n",
    "set_yLim_bool = False  # set y-limits of plots if desired (default: False)\n",
    "#yMin = -0.4 ; yMax = 3.8 # example fixed y-limits (uncomment if desired)\n",
    "\n",
    "libAndData = '_enrichment_N50C_input>=5_noPcounts'\n",
    "\n",
    "variable = 'log2_score'\n",
    "\n",
    "ylabel = 'mean enrichment score'\n",
    "\n",
    "reference_dataframe = N50C_df #aka enrichment_df      \n",
    "### for plotting average value across whole dataset (red horizontal line)\n",
    "\n",
    "specialNote = ''   #(none, = default)\n",
    "\n",
    "motif = motifName\n",
    "startSite=1\n",
    "\n",
    "enrich_vs_startPos_motif = []\n",
    "SEM_enrich_vs_startPos_motif = []\n",
    "startPos = []\n",
    "count_at_pos = []\n",
    "\n",
    "for startSite in range(1,51):\n",
    "    startPos.append(startSite)\n",
    "    mean_score = dataframe[(dataframe['motif_id']==motif)\\\n",
    "                                    &(dataframe['start']==startSite)][variable].mean()\n",
    "    SEM_score = dataframe[(dataframe['motif_id']==motif)\\\n",
    "                                &(dataframe['start']==startSite)][variable].sem()\n",
    "    count = dataframe[(dataframe['motif_id']==motif)\\\n",
    "                                &(dataframe['start']==startSite)][variable].shape[0]\n",
    "\n",
    "    SEM_enrich_vs_startPos_motif.append(SEM_score)\n",
    "\n",
    "    enrich_vs_startPos_motif.append(mean_score)\n",
    "\n",
    "    count_at_pos.append(count)\n",
    "\n",
    "\n",
    "plt.errorbar(startPos, enrich_vs_startPos_motif, yerr=SEM_enrich_vs_startPos_motif, ecolor='grey',\\\n",
    "             fmt='b.-', capsize=2, capthick=1.5) \n",
    "\n",
    "\n",
    "title = 'Motif ' + motif + ' ' + specialNote + ', N50C'  \n",
    "\n",
    "plt.title(title)\n",
    "plt.xlabel('start position')\n",
    "plt.ylabel(ylabel)\n",
    "if(set_yLim_bool==True):\n",
    "    plt.ylim(yMin,yMax)\n",
    "plt.axhline(reference_dataframe[variable].mean(),color='red')\n",
    "\n",
    "\n",
    "figName = 'motif_'+motifName+' '+specialNote+libAndData\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "print count_at_pos\n",
    "print np.mean(count_at_pos)\n",
    "\n",
    "\n",
    "counter = 1\n",
    "enrich_vs_startPos_SUM = []\n",
    "SEM_enrich_vs_startPos_SUMofSQUARES = []\n",
    "\n",
    "for i in range(1,end):\n",
    "    \n",
    "    motif = motifName + '_shuffle' + str(counter) \n",
    "    startSite = 1\n",
    "    \n",
    "    enrich_vs_startPos = []\n",
    "    SEM_enrich_vs_startPos = []\n",
    "   \n",
    "    for startSite in range(1,51):\n",
    "        #startPos.append(startSite)\n",
    "        mean_score = dataframe[(dataframe['motif_id']==motif)\\\n",
    "                                        &(dataframe['start']==startSite)][variable].mean()\n",
    "        SEM_score = dataframe[(dataframe['motif_id']==motif)\\\n",
    "                                    &(dataframe['start']==startSite)][variable].sem()\n",
    "        count = dataframe[(dataframe['motif_id']==motif)\\\n",
    "                                    &(dataframe['start']==startSite)][variable].shape[0]\n",
    "\n",
    "        SEM_enrich_vs_startPos.append(SEM_score)\n",
    "        \n",
    "        enrich_vs_startPos.append(mean_score)\n",
    "        \n",
    "        #count_at_pos.append(count)\n",
    "        \n",
    "    if i == 1:\n",
    "        enrich_vs_startPos_SUM.append(enrich_vs_startPos)\n",
    "        SEM_enrich_vs_startPos_Squared = np.square(SEM_enrich_vs_startPos)\n",
    "        SEM_enrich_vs_startPos_SUMofSQUARES.append(SEM_enrich_vs_startPos_Squared)\n",
    "    else:\n",
    "        enrich_vs_startPos_SUM = np.add(enrich_vs_startPos_SUM, enrich_vs_startPos)\n",
    "        SEM_enrich_vs_startPos_Squared = np.square(SEM_enrich_vs_startPos)\n",
    "        SEM_enrich_vs_startPos_SUMofSQUARES = np.add(SEM_enrich_vs_startPos_SUMofSQUARES,\\\n",
    "                                                     SEM_enrich_vs_startPos_Squared)\n",
    "\n",
    "    #print enrich_vs_startPos    \n",
    "    counter+=1\n",
    "\n",
    "if lastMotif > -1:\n",
    "    #print enrich_vs_startPos_SUM\n",
    "    #print str(float(lastMotif))\n",
    "    enrich_vs_startPos_shuffleAVG = enrich_vs_startPos_SUM/float(lastMotif)\n",
    "    enrich_vs_startPos_shuffleAVG = enrich_vs_startPos_shuffleAVG.tolist()\n",
    "    enrich_vs_startPos_shuffleAVG = enrich_vs_startPos_shuffleAVG[0]\n",
    "    #print enrich_vs_startPos_shuffleAVG\n",
    "    SEM_enrich_vs_startPos_shuffleAVG = (1.0/float(lastMotif)) * np.sqrt(SEM_enrich_vs_startPos_SUMofSQUARES)\n",
    "    SEM_enrich_vs_startPos_shuffleAVG = SEM_enrich_vs_startPos_shuffleAVG.tolist()\n",
    "    SEM_enrich_vs_startPos_shuffleAVG = SEM_enrich_vs_startPos_shuffleAVG[0]\n",
    "    #print SEM_enrich_vs_startPos_shuffleAVG\n",
    "\n",
    "\n",
    "    plt.errorbar(startPos, enrich_vs_startPos_shuffleAVG, yerr=SEM_enrich_vs_startPos_shuffleAVG, ecolor='grey',\\\n",
    "                 fmt='g.-', capsize=2, capthick=1.5) #new! in progress\n",
    "\n",
    "   \n",
    "    title = 'Motif ' + motifName + ' ' + specialNote + '_avg(shuffles), N50C' \n",
    "\n",
    "\n",
    "    plt.title(title)\n",
    "    plt.xlabel('start position')\n",
    "    plt.ylabel(ylabel)\n",
    "    if(set_yLim_bool==True):\n",
    "        plt.ylim(yMin,yMax)\n",
    "    plt.axhline(reference_dataframe[variable].mean(),color='red')\n",
    "\n",
    "    \n",
    "    figName = 'motif_'+motifName+' '+specialNote+'_avg(shuffles)'+libAndData \n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "\n",
    "    plt.errorbar(startPos, enrich_vs_startPos_motif, yerr=SEM_enrich_vs_startPos_motif, ecolor='grey',\\\n",
    "                 fmt='b.-', capsize=2, capthick=1.5)\n",
    "    \n",
    "    plt.errorbar(startPos, enrich_vs_startPos_shuffleAVG, yerr=SEM_enrich_vs_startPos_shuffleAVG, ecolor='grey',\\\n",
    "                 fmt='g.-', capsize=2, capthick=1.5)\n",
    "\n",
    "    title = 'Motif ' + motifName + ' ' + specialNote + '_VS_avg(shuffled), N50C'\n",
    "\n",
    "    plt.title(title)\n",
    "    plt.xlabel('start position')\n",
    "    plt.ylabel(ylabel)\n",
    "    if(set_yLim_bool==True):\n",
    "        plt.ylim(yMin,yMax)\n",
    "    plt.axhline(reference_dataframe[variable].mean(),color='red')\n",
    "\n",
    "    \n",
    "    figName = 'motif_'+motifName+' '+specialNote+'_VS_avg(shuffled)'+libAndData \n",
    "\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Comprehensive analysis of average effects of k-mer sequence motifs (including for 6-mers) ###\n",
    "\n",
    "## select dataset you want to perform the k-mer analysis on:\n",
    "#v1: comprehensive 6-mer results for growth selection enrichment, N50C lib, 5 input reads min, no pseudocounts\n",
    "analysisDF = N50C_df\n",
    "\n",
    "##v2: comprehensive 6-mer results for growth selection enrichment, N50EPC lib, 5 input reads min, no pseudocounts\n",
    "# analysisDF = N50EPC_enr_df\n",
    "\n",
    "\n",
    "## first, make kmers list for desired kmer length:\n",
    "\n",
    "alphabet = ['A','C','G','T']\n",
    "kmer_len = 4  ### #4 #5 #6 #7 #8\n",
    "kmers_list = [''.join(i) for i in itertools.product(alphabet, repeat=kmer_len)]\n",
    "\n",
    "def sem(df):\n",
    "    return np.std(df) / np.sqrt(df.shape[0])\n",
    "\n",
    "#print kmers_list\n",
    "\n",
    "mean_kmer_effect_list = []\n",
    "sem_kmer_effect_list = []\n",
    "N_kmer_effect_list = []\n",
    "mean_NOkmer_effect_list = []\n",
    "sem_NOkmer_effect_list = []\n",
    "N_NOkmer_effect_list = []\n",
    "\n",
    "## next, calculate average effect of kmer of each length in the desired dataset:\n",
    "\n",
    "for kmer in kmers_list:\n",
    "    #print kmer\n",
    "    kmerDF = analysisDF[analysisDF['UTR'].str.contains(kmer)]\n",
    "    NOkmerDF = analysisDF[analysisDF['UTR'].str.contains(kmer)==False]\n",
    "    #print np.mean(kmerDF['log2_score'])\n",
    "    #print np.mean(NOkmerDF['log2_score'])\n",
    "    #print\n",
    "    mean_kmer_effect_list.append(np.mean(kmerDF['log2_score']))\n",
    "    sem_kmer_effect_list.append(sem(kmerDF['log2_score']))\n",
    "    N_kmer_effect_list.append(kmerDF['log2_score'].shape[0])\n",
    "    #\n",
    "    mean_NOkmer_effect_list.append(np.mean(NOkmerDF['log2_score']))\n",
    "    sem_NOkmer_effect_list.append(sem(NOkmerDF['log2_score']))\n",
    "    N_NOkmer_effect_list.append(NOkmerDF['log2_score'].shape[0])\n",
    "    \n",
    "kmer_results_df = pandas.DataFrame()\n",
    "kmer_results_df['kmer_sequence'] = kmers_list\n",
    "kmer_results_df['kmer_enr_mean'] = mean_kmer_effect_list\n",
    "kmer_results_df['kmer_enr_sem'] =  sem_kmer_effect_list\n",
    "kmer_results_df['kmer_enr_N'] =  N_kmer_effect_list\n",
    "kmer_results_df['NOkmer_enr_mean'] = mean_NOkmer_effect_list\n",
    "kmer_results_df['NOkmer_enr_sem'] =  sem_NOkmer_effect_list\n",
    "kmer_results_df['NOkmer_enr_N'] =  N_NOkmer_effect_list\n",
    "\n",
    "kmer_results_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_results_df['kmer_ID'] = kmer_results_df.index + 1\n",
    "kmer_results_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysisDF = kmer_results_df.sort_values(by='kmer_enr_mean',ascending=False)\n",
    "analysisDF['kmer_rank'] = np.arange(len(analysisDF)) + 1\n",
    "analysisDF.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## export the results:\n",
    "### V1: k-mers from N50C lib (5 input reads min, no Pcounts)\n",
    "'''\n",
    "#v1.1: comprehensive 4-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all4mers_Enr_N50C\" + \".csv\"\n",
    "kmer_results_df.to_csv(df_path)\n",
    "#v1.1B: SORTED comprehensive 4-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all4mers_Enr_N50C\" + \".csv\"\n",
    "analysisDF.to_csv(df_path)\n",
    "'''\n",
    "\n",
    "'''\n",
    "#v1.2: comprehensive 5-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all5mers_Enr_N50C\" + \".csv\"\n",
    "kmer_results_df.to_csv(df_path)\n",
    "#v1.2B: SORTED comprehensive 5-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all5mers_Enr_N50C\" + \".csv\"\n",
    "analysisDF.to_csv(df_path)\n",
    "'''\n",
    "\n",
    "'''\n",
    "#v1.3: comprehensive 6-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all6mers_Enr_N50C\" + \".csv\"\n",
    "kmer_results_df.to_csv(df_path)\n",
    "#v1.3B: SORTED comprehensive 6-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all6mers_Enr_N50C\" + \".csv\"\n",
    "analysisDF.to_csv(df_path)\n",
    "'''\n",
    "\n",
    "'''\n",
    "#v1.4: comprehensive 7-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all7mers_Enr_N50C\" + \".csv\"\n",
    "kmer_results_df.to_csv(df_path)\n",
    "#v1.4B: SORTED comprehensive 7-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all7mers_Enr_N50C\" + \".csv\"\n",
    "analysisDF.to_csv(df_path)\n",
    "'''\n",
    "\n",
    "'''\n",
    "#v1.5: comprehensive 8-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all8mers_Enr_N50C\" + \".csv\"\n",
    "kmer_results_df.to_csv(df_path)\n",
    "#v1.5B: SORTED comprehensive 8-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all8mers_Enr_N50C\" + \".csv\"\n",
    "analysisDF.to_csv(df_path)\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "###V2: 6-mers from N50EPC lib, 5 input reads min, no PCounts\n",
    "'''\n",
    "#v2: comprehensive 6-mer results for growth selection enrichment, N50EPC lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all6mers_Enr_N50EPC\" + \".csv\"\n",
    "kmer_results_df.to_csv(df_path)\n",
    "#v2B: SORTED comprehensive 6-mer results for growth selection enrichment, N50EPC lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all6mers_Enr_N50EPC\" + \".csv\"\n",
    "analysisDF.to_csv(df_path)\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load (import) previously exported results: (used in subsequent steps, see below)\n",
    "\n",
    "### V1: k-mers from N50C lib (5 input reads min, no Pcounts)\n",
    "#v1: comprehensive 6-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all6mers_Enr_N50C\" + \".csv\"\n",
    "kmer_results_df = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "#v1B: SORTED comprehensive 6-mer results for growth selection enrichment, N50C lib, 5 input reads min, no PCounts\n",
    "df_path = \"/[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all6mers_Enr_N50C\" + \".csv\"\n",
    "analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "\n",
    "### (or etc. as above for 4-8 mers, as appropriate)\n",
    "\n",
    "\n",
    "\n",
    "### V2: 6-mers from N50EPC lib, 5 input reads min, no PCounts\n",
    "'''\n",
    "#load v2: comprehensive 6-mer results for growth selection enrichment, N50EPC lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"kmer_results_all6mers_Enr_N50EPC\" + \".csv\"\n",
    "kmer_results_df = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "#load v2B: SORTED comprehensive 6-mer results for growth selection enrichment, N50EPC lib, 5 input reads min, no PCounts\n",
    "df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "+ \"Sorted_kmer_results_all6mers_Enr_N50EPC\" + \".csv\"\n",
    "analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot k-mer analysis results\n",
    "\n",
    "wtCYC1_Enr_N50C = N50C_df[N50C_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "wtCYC1_Enr_N50EPC = N50EPC_enr_df[N50EPC_enr_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "\n",
    "allData_df_list = ['N50C_df', 'N50EPC_enr_df', 'N50EPC_enr_df_zoomIn']\n",
    "\n",
    "for allData_df_name in allData_df_list :\n",
    "    \n",
    "    if (allData_df_name == 'N50C_df'):\n",
    "        allData_df = N50C_df\n",
    "        figName = \"kmer_results_all6mers_Enr__from_N50C\" \n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all6mers_Enr_N50C\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "        CYC1_Enr = wtCYC1_Enr_N50C\n",
    "        \n",
    "    elif ((allData_df_name == 'N50EPC_enr_df') | (allData_df_name == 'N50EPC_enr_df_zoomIn')):\n",
    "        allData_df = N50EPC_enr_df\n",
    "        figName = \"kmer_results_all6mers_Enr__from_N50EPC\"\n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all6mers_Enr_N50EPC\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "        CYC1_Enr = wtCYC1_Enr_N50EPC\n",
    "    \n",
    "    if (allData_df_name == 'N50EPC_enr_df_zoomIn'):\n",
    "        figName = \"kmer_results_all6mers_Enr__from_N50EPC_zoomedIn\"\n",
    "    \n",
    "    analysisDF['kmer_rank'] = np.arange(len(analysisDF)) + 1\n",
    "\n",
    "    kmerRank = analysisDF['kmer_rank'].tolist()\n",
    "\n",
    "    kmer_mean = analysisDF['kmer_enr_mean'].tolist()\n",
    "\n",
    "    kmer_sem = analysisDF['kmer_enr_sem'].tolist()\n",
    "\n",
    "    NOkmer_mean = analysisDF['NOkmer_enr_mean'].tolist()\n",
    "\n",
    "    NOkmer_sem = analysisDF['NOkmer_enr_sem'].tolist()\n",
    "\n",
    "    plt.errorbar(kmerRank, kmer_mean, yerr = kmer_sem, capsize=3, alpha=1, fmt='.', markersize = 5, elinewidth = 1)\n",
    "    plt.errorbar(kmerRank, NOkmer_mean, yerr = NOkmer_sem, capsize=3, color ='green', alpha=1, fmt='.',\\\n",
    "                 markersize = 5, elinewidth = 1)\n",
    "    plt.hlines(y=np.mean(allData_df['log2_score']),xmin=-100+np.min(analysisDF['kmer_rank'])\\\n",
    "               ,xmax=1.1*np.max(analysisDF['kmer_rank']),color='red',alpha=1,linewidth=2)\n",
    "    if(allData_df_name != 'N50EPC_enr_df_zoomIn') :\n",
    "        plt.hlines(y=CYC1_Enr,xmin=-100+np.min(analysisDF['kmer_rank'])\\\n",
    "                   ,xmax=1.1*np.max(analysisDF['kmer_rank']),color='orange',alpha=1,linewidth=2)\n",
    "    plt.xlabel('kmer_rank')\n",
    "    plt.ylabel('mean enrichment of library seqs +- kmer (+-sem)')\n",
    "    plt.title('Mean enrichment of library seqs containing kmer (blue)\\nor lacking kmer (green); '+\\\n",
    "              'red = <Enr>all; orange = CYC1 Enr')\n",
    "    #plt.xlim(0,2)\n",
    "    if(allData_df_name != 'N50EPC_enr_df_zoomIn') :\n",
    "        plt.ylim(-1.25,2.75)  ## for matching N50C, N50EPC library scales\n",
    "\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot results for 4-to-8-mer k-mer analyses\n",
    "\n",
    "kmer_n_list = [4, 5, 6, 7, 8]    # set of kmer effects to plot\n",
    "ymin = -1.1; ymax = 3.1          # setting y-scale for plots\n",
    "\n",
    "wtCYC1_Enr_N50C = N50C_df[N50C_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "wtCYC1_Enr_N50EPC = N50EPC_enr_df[N50EPC_enr_df['UTR']==CYC1_wt_N50]['log2_score'].iloc[0]\n",
    "\n",
    "for kmer_n in kmer_n_list :\n",
    "    CYC1_Enr = wtCYC1_Enr_N50C\n",
    "    # comprehensive k-mer results for growth selection enrichment, N50C lib, 5 input reads min, no pseudocounts:\n",
    "    if kmer_n == 4 :\n",
    "        figName = \"kmer_results_all4mers_Enr__from_N50C\"\n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all4mers_Enr_N50C\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "    if kmer_n == 5 :\n",
    "        figName = \"kmer_results_all5mers_Enr__from_N50C\"\n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all5mers_Enr_N50C\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "    if kmer_n == 6 :\n",
    "        figName = \"kmer_results_all6mers_Enr__from_N50C\"\n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all6mers_Enr_N50C\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "    if kmer_n == 7 :\n",
    "        figName = \"kmer_results_all7mers_Enr__from_N50C\"\n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all7mers_Enr_N50C\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "    if kmer_n == 8 :\n",
    "        figName = \"kmer_results_all8mers_Enr__from_N50C\"\n",
    "        df_path = \"[[KMER_ANALYSIS_DIR]]/\"\\\n",
    "        + \"Sorted_kmer_results_all8mers_Enr_N50C\" + \".csv\"\n",
    "        analysisDF = pandas.DataFrame.from_csv(df_path, index_col=None)\n",
    "\n",
    "    allData_df = enrichment_df\n",
    "\n",
    "    #############################\n",
    "\n",
    "    analysisDF['kmer_rank'] = np.arange(len(analysisDF)) + 1\n",
    "\n",
    "    kmerRank = analysisDF['kmer_rank'].tolist()\n",
    "\n",
    "    kmer_mean = analysisDF['kmer_enr_mean'].tolist()\n",
    "\n",
    "    kmer_sem = analysisDF['kmer_enr_sem'].tolist()\n",
    "\n",
    "    NOkmer_mean = analysisDF['NOkmer_enr_mean'].tolist()\n",
    "\n",
    "    NOkmer_sem = analysisDF['NOkmer_enr_sem'].tolist()\n",
    "\n",
    "    plt.errorbar(kmerRank, kmer_mean, yerr = kmer_sem, capsize=3, alpha=1, fmt='.', markersize = 5, elinewidth = 1)\n",
    "    plt.errorbar(kmerRank, NOkmer_mean, yerr = NOkmer_sem, capsize=3, color ='green', alpha=1, fmt='.',\\\n",
    "                 markersize = 5, elinewidth = 1)\n",
    "    plt.hlines(y=np.mean(allData_df['log2_score']),xmin=-100+np.min(analysisDF['kmer_rank'])\\\n",
    "               ,xmax=1.1*np.max(analysisDF['kmer_rank']),color='red',alpha=1,linewidth=2)\n",
    "    plt.hlines(y=CYC1_Enr,xmin=-100+np.min(analysisDF['kmer_rank'])\\\n",
    "               ,xmax=1.1*np.max(analysisDF['kmer_rank']),color='orange',alpha=1,linewidth=2)\n",
    "    plt.xlabel('kmer_rank')\n",
    "    plt.ylabel('mean enrichment of library seqs +- ' +str(kmer_n)+ 'mer (+-sem)')\n",
    "    plt.title('Mean enrichment of library seqs containing ' +str(kmer_n)+\\\n",
    "              'mer (blue)\\nor lacking ' +str(kmer_n)+ 'mer (green); red = <Enr>all; orange = CYC1 Enr')\n",
    "    \n",
    "    if(kmer_n==4):\n",
    "        plt.xlim(-25,275)\n",
    "    plt.ylim(ymin,ymax)  ## for matching scales across all plots\n",
    "\n",
    "    plt.tight_layout()\n",
    "\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Generate histograms of enrichment (Enr) for 3' UTRs +/- each motif, N50C library \n",
    "## PLUS overlapping histograms of 3' UTR variants containing shuffles/control sequences for each motif.\n",
    "## Also generate corresponding motif statistics dataframe.\n",
    "\n",
    "import re\n",
    "\n",
    "def sem(df):\n",
    "    return np.std(df) / np.sqrt(df.shape[0])\n",
    "\n",
    "def print_parameters(analysisDF,name):\n",
    "    print\n",
    "    print name + ':'\n",
    "    print 'N = ' + str(analysisDF.shape[0])\n",
    "    print 'mean +/- sem = ' + str(np.mean(analysisDF['log2_score'])) + ' +\\- ' + str(sem(analysisDF['log2_score']))\n",
    "    \n",
    "def make_shuffles(motif_input,numShuffles_input):\n",
    "    \n",
    "    motif = motif_input\n",
    "    numShuffles = numShuffles_input\n",
    "    \n",
    "    minDist = len(motif)/2\n",
    "    print minDist\n",
    "\n",
    "    shuffleList = []\n",
    "    shuffleList.append(motif)\n",
    "    #print shuffleList\n",
    "\n",
    "    for i in range (0,numShuffles): #(0,5):\n",
    "        continueLoop = True\n",
    "        while(continueLoop == True):\n",
    "            shuffleX = motif\n",
    "            #print shuffleX\n",
    "            shuffleX_list = list(shuffleX)\n",
    "            random.shuffle(shuffleX_list)\n",
    "            shuffleX = ''.join(shuffleX_list)\n",
    "            #print shuffleX\n",
    "            dist = lev.hamming(shuffleX, motif)\n",
    "            if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "                #print \"success\"\n",
    "                shuffleList.append(shuffleX)\n",
    "                continueLoop = False\n",
    "            #else:\n",
    "                #print \"fail, try again\"\n",
    "\n",
    "    print;print; print shuffleList\n",
    "    return shuffleList\n",
    "    \n",
    "#set up histogram bins\n",
    "bins = np.linspace(-10,10,40)\n",
    "\n",
    "plotDF = N50C_df\n",
    "\n",
    "print_parameters(plotDF, 'N50C, ALL')\n",
    "analysisN = plotDF['log2_score'].shape[0]\n",
    "plt.hist(plotDF['log2_score'], bins, alpha=0.5, color = 'grey', edgecolor='black',\\\n",
    "    density=True, label='N50-C, all variants'+ ',\\nN = ' + str(analysisN) )\n",
    "plt.legend(loc='upper left',prop={'size':7})\n",
    "plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "#plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "plt.tight_layout()\n",
    "\n",
    "#uncomment the below and add appropriate file directory in order to save figures:\n",
    "'''\n",
    "figName = 'N50C_+vs-motif-hists_'+'ALL variants'+'_log2E_hist'\n",
    "output_path = \"[[FIGURES_OUTPUT_DIR]]/\"\\\n",
    "+ figName + '.png'\n",
    "plt.savefig(output_path, dpi=1200)\n",
    "output_path = \"[[FIGURES_OUTPUT_DIR]]/\"\\\n",
    "+ figName + '.pdf'\n",
    "plt.savefig(output_path, dpi=1200)\n",
    "'''\n",
    "\n",
    "plt.show()\n",
    "\n",
    "motifs_list = ['TATATA', 'TTTTTATA', 'GCGCGC', 'AATAAA|AAAAAA', 'TAAT...TAAT', 'TGTA.ATA', 'TGTA.A.TA', 'TGTA....TA',\\\n",
    "              'TTGT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT',\\\n",
    "               'AATATA', 'TTTATA', 'TAAATA', 'TATTTA', 'TATAAA', 'TATATT',\\\n",
    "              'CATATA', 'TGTATA', 'TACATA', 'TATGTA', 'TATACA', 'TATATG',\\\n",
    "              'ATTTTATA','TATTTATA','TTATTATA','TTTATATA','TTTTAATA','TTTTTTTA','TTTTTAAA','TTTTTATT',\\\n",
    "               'ATATAT','TTATAT','AAATAT','ATTTAT','ATAAAT','ATATTT','ATATAA',\\\n",
    "               'TATATATA','AATATATA','TTTATATA','TAAATATA','TATTTATA','TATAAATA','TATATTTA','TATATAAA','TATATATT',\\\n",
    "               'TATATATATA','AATATATATA','TTTATATATA','TAAATATATA','TATTTATATA','TATAAATATA',\\\n",
    "               'TATATTTATA','TATATAAATA','TATATATTTA','TATATATAAA','TATATATATT']\n",
    "\n",
    "motif_names_list = ['(UA)3 EE', 'U5AUA EE', '(GC)3 control', 'AAWAAA PE', 'Puf1&2_site', 'Puf3_site', 'Puf4_site',\\\n",
    "                   'Puf5_site', 'Puf6_site', 'polyU_8X_control1(U4A4)', 'polyU_8X_control2(A4U4)',\\\n",
    "                    'polyU_8X_control3(A8)', 'polyU_8X_control4(U2A3U3)', 'polyU_8X_first25nt', 'polyU_8X_last25nt',\\\n",
    "                   'polyU_8X_control5(AU%)', 'AATATA', 'TTTATA', 'TAAATA', 'TATTTA', 'TATAAA', 'TATATT',\\\n",
    "              'CATATA', 'TGTATA', 'TACATA', 'TATGTA', 'TATACA', 'TATATG',\\\n",
    "              'ATTTTATA','TATTTATA','TTATTATA','TTTATATA','TTTTAATA','TTTTTTTA','TTTTTAAA','TTTTTATT',\\\n",
    "                   'ATATAT','TTATAT','AAATAT','ATTTAT','ATAAAT','ATATTT','ATATAA',\\\n",
    "               'TATATATA','AATATATA','TTTATATA','TAAATATA','TATTTATA','TATAAATA','TATATTTA','TATATAAA','TATATATT',\\\n",
    "               'TATATATATA','AATATATATA','TTTATATATA','TAAATATATA','TATTTATATA','TATAAATATA',\\\n",
    "               'TATATTTATA','TATATAAATA','TATATATTTA','TATATATAAA','TATATATATT']\n",
    "\n",
    "motif_stats_names = []\n",
    "motif_stats_means = []\n",
    "motif_stats_sems = []\n",
    "motif_stats_N = []\n",
    "motif_stats_tTest_pVal_withMotif_VS_withoutMotif = []\n",
    "motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle = []\n",
    "motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle = []\n",
    "\n",
    "motif_stats_names.append('N50C ALL')\n",
    "motif_stats_means.append(np.mean(plotDF['log2_score']))\n",
    "motif_stats_N.append(plotDF['log2_score'].shape[0])\n",
    "motif_stats_sems.append(sem(plotDF['log2_score']))\n",
    "motif_stats_tTest_pVal_withMotif_VS_withoutMotif.append(np.nan)\n",
    "motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle.append(np.nan)\n",
    "motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle.append(np.nan)\n",
    "\n",
    "\n",
    "for i in range(0,len(motifs_list)) :\n",
    "    plotDF = N50C_df\n",
    "    motif = motifs_list[i]\n",
    "    motif_name = motif_names_list[i]\n",
    "    \n",
    "    if(motif=='TATATA') :\n",
    "        numShuffles = 10\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "\n",
    "    elif(motif=='TTTTTATA') :\n",
    "        numShuffles = 15\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='GCGCGC') :\n",
    "        numShuffles = 10\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='AATAAA|AAAAAA') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAATTT'\n",
    "        \n",
    "    elif(motif=='TAAT...TAAT') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TGTA.ATA') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TGTA.A.TA') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TGTA....TA') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TTGT') :\n",
    "        numShuffles = 3\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control1(U4A4)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'TTTTAAAA'\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control2(A4U4)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAATTTT'\n",
    "             \n",
    "    elif(motif_name=='polyU_8X_control3(A8)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAAAAAA'\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control4(U2A3U3)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'TTAAATTT'\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_first25nt') :\n",
    "        numShuffles = -2\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_last25nt') :\n",
    "        numShuffles = -3\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control5(AU%)') :\n",
    "        numShuffles = -4\n",
    "        \n",
    "    elif(i in range(16,27+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAATTT'\n",
    "        \n",
    "    elif(i in range(28,35+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAATTTT'\n",
    "        \n",
    "    elif(i in range(36,42+1)) :\n",
    "        numShuffles = -5\n",
    "        control_motif = 'AAATTT'\n",
    "        \n",
    "    elif(i in range(43,51+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAATTTT'\n",
    "        \n",
    "    elif(i in range(52,62+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAAATTTTT'\n",
    "        \n",
    "    simple_motif_name = motif_name\n",
    "    if('polyU_8X' in motif_name) :\n",
    "        simple_motif_name = 'polyU_8X'\n",
    "\n",
    "    analysisDF = plotDF[plotDF['UTR'].str.contains(motif)==False]\n",
    "    print_parameters(analysisDF,'N50C, non-' + simple_motif_name + '-containing')\n",
    "    analysisN = analysisDF['log2_score'].shape[0]\n",
    "    plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'blue', edgecolor='black',\\\n",
    "             density=True, label='N50-C *not* cont. ' + simple_motif_name + ',\\nN = '+ str(analysisN))\n",
    "    plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "    plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "    #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "    motif_stats_names.append(simple_motif_name+'_NOmotif')\n",
    "    motif_stats_means.append(np.mean(analysisDF['log2_score']))\n",
    "    motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "    motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "    ptest_DF_noMotif = analysisDF\n",
    "    \n",
    "    if (numShuffles == -5) : \n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains('TATATA')==False)\\\n",
    "                                                & (enrichment_df['UTR'].str.contains(motif))]\n",
    "        print_parameters(analysisDF,'N50C, ' + simple_motif_name +' -containing (no TATATA)')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'red', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + simple_motif_name + ' (no TATATA), N = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_+motif')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score']))\n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_withMotif = analysisDF\n",
    "    else:\n",
    "        analysisDF = plotDF[plotDF['UTR'].str.contains(motif)]\n",
    "        print_parameters(analysisDF,'N50C, ' + simple_motif_name +' -containing')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'red', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + simple_motif_name + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_+motif')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score']))\n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_withMotif = analysisDF\n",
    "    \n",
    "    if (numShuffles == -1) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                                                & (enrichment_df['UTR'].str.contains(control_motif))]\n",
    "        print_parameters(analysisDF,'N50C, ' + control_motif +' control')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + control_motif + ' control' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(motif_name+'_control')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score'])) \n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -2) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str[1:25].str.contains(motif))]\n",
    "        print_parameters(analysisDF,'N50C, ' + simple_motif_name +' in first 25 nt')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + simple_motif_name +' in first 25 nt' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_first25nt')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score'])) \n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -3) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str[26:50].str.contains(motif))]\n",
    "        print_parameters(analysisDF,'N50C, ' + simple_motif_name +' in last 25 nt')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + simple_motif_name +' in last 25 nt' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_last25nt')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score'])) \n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -5) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                            & (plotDF['UTR'].str.contains('TATATA')==False)\\\n",
    "                            & (enrichment_df['UTR'].str.contains(control_motif))]\n",
    "        print_parameters(analysisDF,'N50C, ' + control_motif +' control (no TATATA)')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + control_motif + ' control (no TATATA)' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_control')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score'])) \n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -4) :\n",
    "        \n",
    "        ### make T4A4 shuffles: 8-mer %AU control ########\n",
    "        shuffles_motif = 'TTTTAAAA' # T4A4 %AU shuffles control generation (4 U's + 4 A's in various shuffled configurations)\n",
    "        minDist = len(shuffles_motif)/2  ## this is in addition to filter for 'TATA' content (see below)\n",
    "\n",
    "        numShuffles = 35\n",
    "\n",
    "        shuffleList = []\n",
    "        shuffleList.append(shuffles_motif)\n",
    "        #print shuffleList\n",
    "\n",
    "        for i in range (0,numShuffles): #(0,5):\n",
    "            continueLoop = True\n",
    "            while(continueLoop == True):\n",
    "                shuffleX = shuffles_motif\n",
    "                #print shuffleX\n",
    "                shuffleX_list = list(shuffleX)\n",
    "                random.shuffle(shuffleX_list)\n",
    "                shuffleX = ''.join(shuffleX_list)\n",
    "                #print shuffleX\n",
    "                dist = lev.hamming(shuffleX, shuffles_motif)\n",
    "                if ((dist >= minDist) & (shuffleX not in shuffleList) & (('TATA' in shuffleX) == False)) :\n",
    "                    #print \"success\"\n",
    "                    shuffleList.append(shuffleX)\n",
    "                    continueLoop = False\n",
    "                #else:\n",
    "                    #print \"fail, try again\"\n",
    "\n",
    "        print;print; print shuffleList\n",
    "        ##############################################\n",
    "        \n",
    "        shuffles = shuffleList  #want to include T4A4 starting sequence in this case\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                                                & (plotDF['UTR'].str.contains('|'.join(shuffles)))]\n",
    "        print_parameters(analysisDF,'N50C, ' + '%AU (4xA,4xU) control')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + '%AU (4xA,4xU) control' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_%AU(4xA,4xU)control')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score']))\n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    else :\n",
    "        shuffles = shuffleList[1:]  #exclude the consensus motif from the list of shuffled sequences\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                                                & (plotDF['UTR'].str.contains('|'.join(shuffles)))]\n",
    "        print_parameters(analysisDF,'N50C, ' + simple_motif_name +' shuffles')\n",
    "        analysisN = analysisDF['log2_score'].shape[0]\n",
    "        plt.hist(analysisDF['log2_score'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='N50-C cont. ' + simple_motif_name + ' shuffles' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('enrichment', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('N50-C library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_shuffles')\n",
    "        motif_stats_means.append(np.mean(analysisDF['log2_score']))\n",
    "        motif_stats_N.append(analysisDF['log2_score'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['log2_score']))\n",
    "        ptest_DF_control = analysisDF\n",
    "\n",
    "    plt.legend(loc='upper left',prop={'size':7})\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    ### perform T-tests and save p-values #####\n",
    "    Ttest = scipy.stats.ttest_ind(ptest_DF_withMotif['log2_score'],ptest_DF_noMotif['log2_score'],\\\n",
    "                          equal_var=False,nan_policy='propagate')\n",
    "    pVal = Ttest[1]   #;   print pVal \n",
    "    for i in range(0,3) :  #must add to list 3 times because of the table structure (3 entries per motif)\n",
    "        motif_stats_tTest_pVal_withMotif_VS_withoutMotif.append(pVal)\n",
    "\n",
    "    Ttest = scipy.stats.ttest_ind(ptest_DF_withMotif['log2_score'],ptest_DF_control['log2_score'],\\\n",
    "                          equal_var=False,nan_policy='propagate')\n",
    "    pVal = Ttest[1]   #;   print pVal \n",
    "    for i in range(0,3) :  #must add to list 3 times because of the table structure (3 entries per motif)\n",
    "        motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle.append(pVal)\n",
    "    \n",
    "    Ttest = scipy.stats.ttest_ind(ptest_DF_noMotif['log2_score'],ptest_DF_control['log2_score'],\\\n",
    "                          equal_var=False,nan_policy='propagate')\n",
    "    pVal = Ttest[1]   #;   print pVal \n",
    "    for i in range(0,3) :  #must add to list 3 times because of the table structure (3 entries per motif)\n",
    "        motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle.append(pVal)\n",
    "    #############################################\n",
    "    \n",
    "    # uncomment the below to save the plots (add desired file directory)\n",
    "    '''\n",
    "    figName = 'N50C_+vs-motif-hists_'+motif_name+'_log2E_hist'\n",
    "    output_path = \"[[FIGURES_OUTPUT_DIR]]/\"\\\n",
    "    + figName + '.png'\n",
    "    plt.savefig(output_path, dpi=1200)\n",
    "    output_path = \"[[FIGURES_OUTPUT_DIR]]/\"\\\n",
    "    + figName + '.pdf'\n",
    "    plt.savefig(output_path, dpi=1200)\n",
    "    '''\n",
    "    \n",
    "    plt.show()  ## optional, if want to see all plots\n",
    "    #plt.close()  ## alternative, if only want to save all plots\n",
    "    \n",
    "\n",
    "print motif_stats_names; print motif_stats_means; print motif_stats_sems; print motif_stats_N;\\\n",
    "print motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle; print motif_stats_tTest_pVal_withMotif_VS_withoutMotif;\\\n",
    "print motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle\n",
    "\n",
    "print; print len(motif_stats_tTest_pVal_withMotif_VS_withoutMotif);\\\n",
    "print len(motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle);\\\n",
    "print len(motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle)\n",
    "\n",
    "N50C_motif_stats_df = pandas.DataFrame()\n",
    "\n",
    "N50C_motif_stats_df['motif_stats_names'] = motif_stats_names\n",
    "N50C_motif_stats_df['motif_stats_means'] = motif_stats_means\n",
    "N50C_motif_stats_df['motif_stats_sems'] = motif_stats_sems\n",
    "N50C_motif_stats_df['motif_stats_N'] = motif_stats_N\n",
    "N50C_motif_stats_df['motif_stats_tTest_pVal_withMotif_VS_withoutMotif']\\\n",
    "= motif_stats_tTest_pVal_withMotif_VS_withoutMotif\n",
    "N50C_motif_stats_df['motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle']\\\n",
    "= motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle\n",
    "N50C_motif_stats_df['motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle']\\\n",
    "= motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle\n",
    "\n",
    "print N50C_motif_stats_df.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## export motif stats dataframe (for enrichment in N50C library) as CSV file\n",
    "df_path = \"[[OUTPUT_DIR_MOTIF_STATS]]/\"\\\n",
    "+ \"Motif_Enr_Stats_from_N50C\" + \".csv\"\n",
    "N50C_motif_stats_df.to_csv(df_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Generate histograms of mRNA half-lives (Geisberg et al. 2014) for native yeast 3' UTRs +/- each motif\n",
    "## PLUS overlapping histograms of 3' UTR variants containing shuffles/control sequences for each motif.\n",
    "## Also generate corresponding motif statistics dataframe.\n",
    "\n",
    "import re\n",
    "\n",
    "def sem(df):\n",
    "    return np.std(df) / np.sqrt(df.shape[0])\n",
    "\n",
    "def print_parameters(analysisDF,name):\n",
    "    print\n",
    "    print name + ':'\n",
    "    print 'N = ' + str(analysisDF.shape[0])\n",
    "    print 'mean +/- sem = ' + str(np.mean(analysisDF['Geisberg/min'])) + ' +\\- ' + str(sem(analysisDF['Geisberg/min']))\n",
    "    \n",
    "def make_shuffles(motif_input,numShuffles_input):\n",
    "    \n",
    "    motif = motif_input\n",
    "    numShuffles = numShuffles_input\n",
    "    \n",
    "    minDist = len(motif)/2\n",
    "    print minDist\n",
    "\n",
    "    shuffleList = []\n",
    "    shuffleList.append(motif)\n",
    "    #print shuffleList\n",
    "\n",
    "    for i in range (0,numShuffles): #(0,5):\n",
    "        continueLoop = True\n",
    "        while(continueLoop == True):\n",
    "            shuffleX = motif\n",
    "            #print shuffleX\n",
    "            shuffleX_list = list(shuffleX)\n",
    "            random.shuffle(shuffleX_list)\n",
    "            shuffleX = ''.join(shuffleX_list)\n",
    "            #print shuffleX\n",
    "            dist = lev.hamming(shuffleX, motif)\n",
    "            if ((dist >= minDist) & (shuffleX not in shuffleList)) :\n",
    "                #print \"success\"\n",
    "                shuffleList.append(shuffleX)\n",
    "                continueLoop = False\n",
    "            #else:\n",
    "                #print \"fail, try again\"\n",
    "\n",
    "    print;print; print shuffleList\n",
    "    return shuffleList\n",
    "\n",
    "    \n",
    "#set up histogram bins\n",
    "bins = np.linspace(0,150,50)\n",
    "\n",
    "plotDF = nativeUTRs_df\n",
    "\n",
    "print_parameters(plotDF, 'native, ALL')\n",
    "analysisN = plotDF['Geisberg/min'].shape[0]\n",
    "plt.hist(plotDF['Geisberg/min'], bins, alpha=0.5, color = 'grey', edgecolor='black',\\\n",
    "    density=True, label='native, all variants'+ ',\\nN = ' + str(analysisN) )\n",
    "plt.legend(loc='upper right',prop={'size':7})\n",
    "plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "#plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "plt.tight_layout()\n",
    "\n",
    "# uncomment the below to save figures (add desired output filepath)\n",
    "'''\n",
    "figName = 'native_+vs-motif-hists_'+'ALL variants'+'_halfLife(Geisberg)_hist'\n",
    "output_path = \"[[FIGURES_OUTPUT_DIRECTORY]]/\"\\\n",
    "+ figName + '.png'\n",
    "plt.savefig(output_path, dpi=1200)\n",
    "output_path = \"[[FIGURES_OUTPUT_DIRECTORY]]/\"\\\n",
    "+ figName + '.pdf'\n",
    "plt.savefig(output_path, dpi=1200)\n",
    "'''\n",
    "\n",
    "plt.show()\n",
    "\n",
    "motifs_list = ['TATATA', 'TTTTTATA', 'GCGCGC', 'AATAAA|AAAAAA', 'TAAT...TAAT', 'TGTA.ATA', 'TGTA.A.TA', 'TGTA....TA',\\\n",
    "              'TTGT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT', 'TTTTTTTT',\\\n",
    "               'AATATA', 'TTTATA', 'TAAATA', 'TATTTA', 'TATAAA', 'TATATT',\\\n",
    "              'CATATA', 'TGTATA', 'TACATA', 'TATGTA', 'TATACA', 'TATATG',\\\n",
    "              'ATTTTATA','TATTTATA','TTATTATA','TTTATATA','TTTTAATA','TTTTTTTA','TTTTTAAA','TTTTTATT',\\\n",
    "               'ATATAT','TTATAT','AAATAT','ATTTAT','ATAAAT','ATATTT','ATATAA',\\\n",
    "               'TATATATA','AATATATA','TTTATATA','TAAATATA','TATTTATA','TATAAATA','TATATTTA','TATATAAA','TATATATT',\\\n",
    "               'TATATATATA','AATATATATA','TTTATATATA','TAAATATATA','TATTTATATA','TATAAATATA',\\\n",
    "               'TATATTTATA','TATATAAATA','TATATATTTA','TATATATAAA','TATATATATT']\n",
    "\n",
    "motif_names_list = ['(UA)3 EE', 'U5AUA EE', '(GC)3 control', 'AAWAAA PE', 'Puf1&2_site', 'Puf3_site', 'Puf4_site',\\\n",
    "                   'Puf5_site', 'Puf6_site', 'polyU_8X_control1(U4A4)', 'polyU_8X_control2(A4U4)',\\\n",
    "                    'polyU_8X_control3(A8)', 'polyU_8X_control4(U2A3U3)', 'polyU_8X_first25nt', 'polyU_8X_last25nt',\\\n",
    "                   'polyU_8X_control5(AU%)', 'AATATA', 'TTTATA', 'TAAATA', 'TATTTA', 'TATAAA', 'TATATT',\\\n",
    "              'CATATA', 'TGTATA', 'TACATA', 'TATGTA', 'TATACA', 'TATATG',\\\n",
    "              'ATTTTATA','TATTTATA','TTATTATA','TTTATATA','TTTTAATA','TTTTTTTA','TTTTTAAA','TTTTTATT',\\\n",
    "                   'ATATAT','TTATAT','AAATAT','ATTTAT','ATAAAT','ATATTT','ATATAA',\\\n",
    "               'TATATATA','AATATATA','TTTATATA','TAAATATA','TATTTATA','TATAAATA','TATATTTA','TATATAAA','TATATATT',\\\n",
    "               'TATATATATA','AATATATATA','TTTATATATA','TAAATATATA','TATTTATATA','TATAAATATA',\\\n",
    "               'TATATTTATA','TATATAAATA','TATATATTTA','TATATATAAA','TATATATATT']\n",
    "\n",
    "motif_stats_names = []\n",
    "motif_stats_means = []\n",
    "motif_stats_sems = []\n",
    "motif_stats_N = []\n",
    "motif_stats_tTest_pVal_withMotif_VS_withoutMotif = []\n",
    "motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle = []\n",
    "motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle = []\n",
    "\n",
    "motif_stats_names.append('native ALL')\n",
    "motif_stats_means.append(np.mean(plotDF['Geisberg/min']))\n",
    "motif_stats_N.append(plotDF['Geisberg/min'].shape[0])\n",
    "motif_stats_sems.append(sem(plotDF['Geisberg/min']))\n",
    "motif_stats_tTest_pVal_withMotif_VS_withoutMotif.append(np.nan)\n",
    "motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle.append(np.nan)\n",
    "motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle.append(np.nan)\n",
    "\n",
    "\n",
    "for i in range(0,len(motifs_list)) :\n",
    "    plotDF = nativeUTRs_df\n",
    "    motif = motifs_list[i]\n",
    "    motif_name = motif_names_list[i]\n",
    "    \n",
    "    if(motif=='TATATA') :\n",
    "        numShuffles = 10\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "\n",
    "    elif(motif=='TTTTTATA') :\n",
    "        numShuffles = 15\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='GCGCGC') :\n",
    "        numShuffles = 10\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='AATAAA|AAAAAA') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAATTT'\n",
    "        \n",
    "    elif(motif=='TAAT...TAAT') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TGTA.ATA') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TGTA.A.TA') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TGTA....TA') :\n",
    "        numShuffles = 50\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif=='TTGT') :\n",
    "        numShuffles = 3\n",
    "        shuffleList = make_shuffles(motif,numShuffles)\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control1(U4A4)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'TTTTAAAA'\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control2(A4U4)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAATTTT'\n",
    "             \n",
    "    elif(motif_name=='polyU_8X_control3(A8)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAAAAAA'\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control4(U2A3U3)') :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'TTAAATTT'\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_first25nt') :\n",
    "        numShuffles = -2\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_last25nt') :\n",
    "        numShuffles = -3\n",
    "        \n",
    "    elif(motif_name=='polyU_8X_control5(AU%)') :\n",
    "        numShuffles = -4\n",
    "        \n",
    "    elif(i in range(16,27+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAATTT'\n",
    "        \n",
    "    elif(i in range(28,35+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAATTTT'\n",
    "        \n",
    "    elif(i in range(36,42+1)) :\n",
    "        numShuffles = -5\n",
    "        control_motif = 'AAATTT'\n",
    "        \n",
    "    elif(i in range(43,51+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAATTTT'\n",
    "        \n",
    "    elif(i in range(52,62+1)) :\n",
    "        numShuffles = -1\n",
    "        control_motif = 'AAAAATTTTT'\n",
    "        \n",
    "\n",
    "    simple_motif_name = motif_name\n",
    "    if('polyU_8X' in motif_name) :\n",
    "        simple_motif_name = 'polyU_8X'\n",
    "\n",
    "\n",
    "    analysisDF = plotDF[plotDF['UTR'].str.contains(motif)==False]\n",
    "    print_parameters(analysisDF,'native, non-' + simple_motif_name + '-containing')\n",
    "    analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "    plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'blue', edgecolor='black',\\\n",
    "             density=True, label='native *not* cont. ' + simple_motif_name + ',\\nN = '+ str(analysisN))\n",
    "    plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "    plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "    #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "    motif_stats_names.append(simple_motif_name+'_NOmotif')\n",
    "    motif_stats_means.append(np.mean(analysisDF['Geisberg/min']))\n",
    "    motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "    motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "    ptest_DF_noMotif = analysisDF\n",
    "    \n",
    "    if (numShuffles == -5) : \n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains('TATATA')==False)\\\n",
    "                                                & (enrichment_df['UTR'].str.contains(motif))]\n",
    "        print_parameters(analysisDF,'native, ' + simple_motif_name +' -containing (no TATATA)')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'red', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + simple_motif_name + ' (no TATATA), N = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_+motif')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min']))\n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_withMotif = analysisDF\n",
    "    else:\n",
    "        analysisDF = plotDF[plotDF['UTR'].str.contains(motif)]\n",
    "        print_parameters(analysisDF,'native, ' + simple_motif_name +' -containing')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'red', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + simple_motif_name + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_+motif')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min']))\n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_withMotif = analysisDF\n",
    "    \n",
    "    if (numShuffles == -1) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                                                & (enrichment_df['UTR'].str.contains(control_motif))]\n",
    "        print_parameters(analysisDF,'native, ' + control_motif +' control')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + control_motif + ' control' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(motif_name+'_control')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min'])) \n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -2) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str[1:25].str.contains(motif))]\n",
    "        print_parameters(analysisDF,'native, ' + simple_motif_name +' in first 25 nt')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + simple_motif_name +' in first 25 nt' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_first25nt')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min'])) \n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -3) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str[26:50].str.contains(motif))]\n",
    "        print_parameters(analysisDF,'native, ' + simple_motif_name +' in last 25 nt')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + simple_motif_name +' in last 25 nt' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_last25nt')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min'])) \n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -5) :\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                            & (plotDF['UTR'].str.contains('TATATA')==False)\\\n",
    "                            & (enrichment_df['UTR'].str.contains(control_motif))]\n",
    "        print_parameters(analysisDF,'native, ' + control_motif +' control (no TATATA)')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + control_motif + ' control (no TATATA)' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_control')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min'])) \n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    elif (numShuffles == -4) :\n",
    "        \n",
    "        ### make T4A4 shuffles: 8-mer %AU control ########\n",
    "        shuffles_motif = 'TTTTAAAA' # T4A4 %AU shuffles control generation (4 U's + 4 A's in various shuffled configurations)\n",
    "        minDist = len(shuffles_motif)/2  ## this is in addition to filter for 'TATA' content (see below)\n",
    "\n",
    "        numShuffles = 35\n",
    "\n",
    "        shuffleList = []\n",
    "        shuffleList.append(shuffles_motif)\n",
    "        #print shuffleList\n",
    "\n",
    "        for i in range (0,numShuffles): #(0,5):\n",
    "            continueLoop = True\n",
    "            while(continueLoop == True):\n",
    "                shuffleX = shuffles_motif\n",
    "                #print shuffleX\n",
    "                shuffleX_list = list(shuffleX)\n",
    "                random.shuffle(shuffleX_list)\n",
    "                shuffleX = ''.join(shuffleX_list)\n",
    "                #print shuffleX\n",
    "                dist = lev.hamming(shuffleX, shuffles_motif)\n",
    "                if ((dist >= minDist) & (shuffleX not in shuffleList) & (('TATA' in shuffleX) == False)) :\n",
    "                    #print \"success\"\n",
    "                    shuffleList.append(shuffleX)\n",
    "                    continueLoop = False\n",
    "                #else:\n",
    "                    #print \"fail, try again\"\n",
    "\n",
    "        print;print; print shuffleList\n",
    "        ##############################################\n",
    "        \n",
    "        shuffles = shuffleList  #want to include T4A4 starting sequence in this case\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                                                & (plotDF['UTR'].str.contains('|'.join(shuffles)))]\n",
    "        print_parameters(analysisDF,'native, ' + '%AU (4xA,4xU) control')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + '%AU (4xA,4xU) control' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_%AU(4xA,4xU)control')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min']))\n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_control = analysisDF\n",
    "    else :\n",
    "        shuffles = shuffleList[1:]  #exclude the consensus motif from the list of shuffled sequences\n",
    "        analysisDF = plotDF[(plotDF['UTR'].str.contains(motif)==False)\\\n",
    "                                                & (plotDF['UTR'].str.contains('|'.join(shuffles)))]\n",
    "        print_parameters(analysisDF,'native, ' + simple_motif_name +' shuffles')\n",
    "        analysisN = analysisDF['Geisberg/min'].shape[0]\n",
    "        plt.hist(analysisDF['Geisberg/min'], bins, alpha=0.5, color = 'white', edgecolor='black',\\\n",
    "                 density=True, label='native cont. ' + simple_motif_name + ' shuffles' + ',\\nN = '+ str(analysisN))\n",
    "        plt.xlabel('half-life (1/min)', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        plt.ylabel('density', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        #plt.title('native library', fontweight='bold', fontsize=14,labelpad=10)\n",
    "        motif_stats_names.append(simple_motif_name+'_shuffles')\n",
    "        motif_stats_means.append(np.mean(analysisDF['Geisberg/min']))\n",
    "        motif_stats_N.append(analysisDF['Geisberg/min'].shape[0])\n",
    "        motif_stats_sems.append(sem(analysisDF['Geisberg/min']))\n",
    "        ptest_DF_control = analysisDF\n",
    "\n",
    "    plt.legend(loc='upper right',prop={'size':7})\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    ### perform T-tests and save p-values #####\n",
    "    Ttest = scipy.stats.ttest_ind(ptest_DF_withMotif['Geisberg/min'],ptest_DF_noMotif['Geisberg/min'],\\\n",
    "                          equal_var=False,nan_policy='propagate')\n",
    "    pVal = Ttest[1]   #;   print pVal \n",
    "    for i in range(0,3) :  #must add to list 3 times because of the table structure (3 entries per motif)\n",
    "        motif_stats_tTest_pVal_withMotif_VS_withoutMotif.append(pVal)\n",
    "\n",
    "    Ttest = scipy.stats.ttest_ind(ptest_DF_withMotif['Geisberg/min'],ptest_DF_control['Geisberg/min'],\\\n",
    "                          equal_var=False,nan_policy='propagate')\n",
    "    pVal = Ttest[1]   #;   print pVal \n",
    "    for i in range(0,3) :  #must add to list 3 times because of the table structure (3 entries per motif)\n",
    "        motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle.append(pVal)\n",
    "    \n",
    "    Ttest = scipy.stats.ttest_ind(ptest_DF_noMotif['Geisberg/min'],ptest_DF_control['Geisberg/min'],\\\n",
    "                          equal_var=False,nan_policy='propagate')\n",
    "    pVal = Ttest[1]   #;   print pVal \n",
    "    for i in range(0,3) :  #must add to list 3 times because of the table structure (3 entries per motif)\n",
    "        motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle.append(pVal)\n",
    "    #############################################\n",
    "    \n",
    "    # uncomment the below to save figures (add desired output filepath)\n",
    "    '''\n",
    "    figName = 'native_+vs-motif-hists_'+motif_name+'_halfLife(Geisberg)_hist'\n",
    "    output_path = \"[[FIGURES_OUTPUT_DIRECTORY]]/\"\\\n",
    "    + figName + '.png'\n",
    "    plt.savefig(output_path, dpi=1200)\n",
    "    output_path = \"[[FIGURES_OUTPUT_DIRECTORY]]/\"\\\n",
    "    + figName + '.pdf'\n",
    "    plt.savefig(output_path, dpi=1200)\n",
    "    '''\n",
    "    \n",
    "    plt.show()  ## optional, if want to see all plots\n",
    "    #plt.close()  ## alternative, if only want to save all plots\n",
    "    \n",
    "\n",
    "print motif_stats_names; print motif_stats_means; print motif_stats_sems; print motif_stats_N;\\\n",
    "print motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle; print motif_stats_tTest_pVal_withMotif_VS_withoutMotif;\\\n",
    "print motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle\n",
    "\n",
    "print; print len(motif_stats_tTest_pVal_withMotif_VS_withoutMotif);\\\n",
    "print len(motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle);\\\n",
    "print len(motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle)\n",
    "\n",
    "native_motif_stats_df = pandas.DataFrame()\n",
    "\n",
    "native_motif_stats_df['motif_stats_names'] = motif_stats_names\n",
    "native_motif_stats_df['motif_stats_means'] = motif_stats_means\n",
    "native_motif_stats_df['motif_stats_sems'] = motif_stats_sems\n",
    "native_motif_stats_df['motif_stats_N'] = motif_stats_N\n",
    "native_motif_stats_df['motif_stats_tTest_pVal_withMotif_VS_withoutMotif']\\\n",
    "= motif_stats_tTest_pVal_withMotif_VS_withoutMotif\n",
    "native_motif_stats_df['motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle']\\\n",
    "= motif_stats_tTest_pVal_withMotif_VS_ControlOrShuffle\n",
    "native_motif_stats_df['motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle']\\\n",
    "= motif_stats_tTest_pVal_withoutMotif_VS_ControlOrShuffle\n",
    "\n",
    "print native_motif_stats_df.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## export motif stats dataframe (for half-lives of native yeast mRNAs, from Geisberg et al. 2014) as CSV file\n",
    "df_path = \"[[MOTIF_STATS_OUTPUT_PATH]]/\"\\\n",
    "+ \"Motif_HalfLife_Stats_from_nativeUTRs_Geisberg\" + \".csv\"\n",
    "native_motif_stats_df.to_csv(df_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate list of native 3' UTR sequences with half-lives\n",
    "\n",
    "## merge native 3' UTR sequence with isoform half-life\n",
    "\n",
    "#Supplementary information from Geisberg et al. \n",
    "RNAs = pandas.read_excel('geisberg_half-lives.xls')\n",
    "isoforms = pandas.read_excel('geisberg_isoforms.xls')\n",
    "\n",
    "#Native sequeneces from curated list of yeast genes with 1000 bp upstream and downstream of ORF from Saccharomyces Genome Database\n",
    "seqs = pandas.read_csv('orf_genomic_1000_all_SGD_names.tsv', sep = '\\t')\n",
    "seqs = seqs.drop_duplicates(subset=['systematic name'], keep=False)\n",
    "\n",
    "\n",
    "## merge dataframes\n",
    "isoforms_seqs = pandas.merge(isoforms, seqs, how = \"inner\", on = 'systematic name')\n",
    "\n",
    "isoforms_seqs[\"Sequence\"] = isoforms_seqs[\"Sequence\"].astype(str)\n",
    "isoforms_seqs['truncated']= isoforms_seqs['Sequence'].str.slice(-1000,)\n",
    "isoforms_seqs[\"Position relative to stop codon\"] = isoforms_seqs[\"Position relative to stop codon\"].fillna(0.0).astype(int)\n",
    "\n",
    "isoforms_seqs['trimmed'] = isoforms_seqs.apply(lambda x: x['truncated'][:x['Position relative to stop codon']], axis=1)\n",
    "\n",
    "del isoforms_seqs['Sequence']\n",
    "del isoforms_seqs['truncated']\n",
    "\n",
    "allseqs = pandas.merge(RNAs, isoforms_seqs, how = 'left', on = 'systematic name')\n",
    "allseqs = allseqs.dropna(subset = ['trimmed'], inplace=True)\n",
    "\n",
    "del allseqs['difference']\n",
    "del allseqs['chrom_x']\n",
    "del allseqs['chrom_y']\n",
    "del allseqs['Description']\n",
    "del allseqs['Absolute Peak Coordinate']\n",
    "del allseqs['SimpleType']\n",
    "del allseqs['common name_y']\n",
    "\n",
    "allseqs.rename(columns={'trimmed': 'sequence', 'common name_x': 'common name', 'Half-Life           (in minutes)':'Gene Half-Life'}).to_csv('Geisberg2014_native-mRNA-half-life_with_sequence.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
